diff --git a/src/condor_tools/preen.cpp b/src/condor_tools/preen.cpp
index d028cc7..3312b35 100644
--- a/src/condor_tools/preen.cpp
+++ b/src/condor_tools/preen.cpp
@@ -137,6 +137,11 @@ main( int argc, char *argv[] )
 				RmFlag = TRUE;
 				break;
 
+              case 'd':
+                  Termlog = 1;
+                  dprintf_config("TOOL");
+                  break;
+
 			  default:
 				usage();
 
