diff --git a/src/condor_birdwatcher/CMakeLists.txt b/src/condor_birdwatcher/CMakeLists.txt
index 1c3a59e..362622a 100644
--- a/src/condor_birdwatcher/CMakeLists.txt
+++ b/src/condor_birdwatcher/CMakeLists.txt
@@ -20,7 +20,7 @@ if (WINDOWS AND NOT WIN_EXEC_NODE_ONLY)
 
 	set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} /SUBSYSTEM:WINDOWS")
 	
-	file( GLOB BWSrcs *.cpp *.h *.rc  res/*)
+	file( GLOB BWSrcs *.cpp *.h *.rc *.rc2 res/*)
 	
 	condor_exe( condor_birdwatcher "${BWSrcs}" ${C_SBIN} "" OFF)
 
diff --git a/src/condor_birdwatcher/birdwatcher.rc b/src/condor_birdwatcher/birdwatcher.rc
index b7d55ba..f559363 100755
--- a/src/condor_birdwatcher/birdwatcher.rc
+++ b/src/condor_birdwatcher/birdwatcher.rc
@@ -7,14 +7,16 @@
 //
 // Generated from the TEXTINCLUDE 2 resource.
 //
-#include "winres.h"
+#include "windows.h"
 
 /////////////////////////////////////////////////////////////////////////////
 #undef APSTUDIO_READONLY_SYMBOLS
 
 /////////////////////////////////////////////////////////////////////////////
 // English (U.S.) resources
-
+#ifndef IDC_STATIC
+#define IDC_STATIC      (-1)
+#endif
 #if !defined(AFX_RESOURCE_DLL) || defined(AFX_TARG_ENU)
 #ifdef _WIN32
 LANGUAGE LANG_ENGLISH, SUBLANG_ENGLISH_US
@@ -34,12 +36,6 @@ END
 
 2 TEXTINCLUDE 
 BEGIN
-    "#include ""winres.h""\r\n"
-    "\0"
-END
-
-3 TEXTINCLUDE 
-BEGIN
     "#define _AFX_NO_SPLITTER_RESOURCES\r\n"
     "#define _AFX_NO_OLE_RESOURCES\r\n"
     "#define _AFX_NO_TRACKER_RESOURCES\r\n"
