diff --git a/src/condor_schedd.V6/schedd.cpp b/src/condor_schedd.V6/schedd.cpp
index 5022171..ca7ef86 100644
--- a/src/condor_schedd.V6/schedd.cpp
+++ b/src/condor_schedd.V6/schedd.cpp
@@ -9261,12 +9261,18 @@ Scheduler::child_exit(int pid, int status)
 	if (IsSchedulerUniverse(srec)) {
  		// scheduler universe process 
 		scheduler_univ_job_exit(pid,status,srec);
+		int status = 0;
+		GetAttributeInt(srec->job_id.cluster, srec->job_id.proc,
+						ATTR_JOB_STATUS, &status);
 		delete_shadow_rec( pid );
 			// even though this will get set correctly in
 			// count_jobs(), try to keep it accurate here, too.  
 		if( SchedUniverseJobsRunning > 0 ) {
 			SchedUniverseJobsRunning--;
 		}
+		if ( status == IDLE ) {
+			SchedUniverseJobsIdle++;
+		}
 	} else if (srec) {
 		char* name = NULL;
 			//
@@ -11152,7 +11158,7 @@ Scheduler::needReschedule()
 		// timeout() has internal logic to avoid doing its work too
 		// frequently, and we want to send the reschedule after
 		// updating our ad in the collector, not before.
-	timeout();
+	daemonCore->Reset_Timer(timeoutid,0,1);
 }
 
 void
