diff --git a/msconfig/condor_birdwatcher.vcproj b/msconfig/condor_birdwatcher.vcproj
index 3c70f9c..5a2cd3b 100644
--- a/msconfig/condor_birdwatcher.vcproj
+++ b/msconfig/condor_birdwatcher.vcproj
@@ -52,6 +52,7 @@
 				AdditionalOptions="$(CONDOR_DEFINE) $(CONDOR_INCLUDE) $(CONDOR_GSOAP_INCLUDE) $(CONDOR_KERB_INCLUDE) $(CONDOR_PCRE_INCLUDE) $(CONDOR_OPENSSL_INCLUDE) $(CONDOR_POSTGRESQL_INCLUDE)"
 				Optimization="1"
 				InlineFunctionExpansion="1"
+				AdditionalIncludeDirectories="%SDK_DIR%\Include\mfc"
 				PreprocessorDefinitions="WIN32;NDEBUG;_WINDOWS"
 				StringPooling="true"
 				RuntimeLibrary="0"
