diff --git a/src/condor_examples/condor_config.generic b/src/condor_examples/condor_config.generic
index 3375e11..e44a38d 100644
--- a/src/condor_examples/condor_config.generic
+++ b/src/condor_examples/condor_config.generic
@@ -1756,14 +1756,6 @@ INVALID_LOG_FILES	= core
 
 JAVA = /usr/bin/java
 
-##  Some JVMs need to be told the maximum amount of heap memory
-##  to offer to the process.  If your JVM supports this, give
-##  the argument here, and Condor will fill in the memory amount.
-##  If left blank, your JVM will choose some default value,
-##  typically 64 MB.  The default (-Xmx) works with the Sun JVM.
-
-JAVA_MAXHEAP_ARGUMENT = -Xmx
-
 ## JAVA_CLASSPATH_DEFAULT gives the default set of paths in which
 ## Java classes are to be found.  Each path is separated by spaces.
 ## If your JVM needs to be informed of additional directories, add
diff --git a/src/condor_utils/java_config.cpp b/src/condor_utils/java_config.cpp
index 70daa66..655658a 100644
--- a/src/condor_utils/java_config.cpp
+++ b/src/condor_utils/java_config.cpp
@@ -41,20 +41,6 @@ int java_config( MyString &cmd, ArgList *args, StringList *extra_classpath )
 	if(!tmp) return 0;
 	cmd = tmp;
 	free(tmp);
-
-	tmp = param("JAVA_MAXHEAP_ARGUMENT");
-	const char *xmx_arg = "";
-	// if tmp is not set or the MAXHEAP argument does not contain the leading -Xmx, add it.
-	if( !tmp || strncmp(tmp, "-Xmx", 4) ) 
-		xmx_arg = "-Xmx";
-
-	if (!tmp) {
-		arg_buf.sprintf("%s%dm",xmx_arg,sysapi_phys_memory()/ sysapi_ncpus());
-	} else {
-		arg_buf.sprintf("%s%s",xmx_arg,tmp);
-		free(tmp);
-	}
-	args->AppendArg(arg_buf.Value());
 	
 	tmp = param("JAVA_CLASSPATH_ARGUMENT");
 	if(!tmp) tmp = strdup("-classpath");
