diff --git a/src/condor_submit.V6/submit.cpp b/src/condor_submit.V6/submit.cpp
index dadfb8b..d3392ad 100644
--- a/src/condor_submit.V6/submit.cpp
+++ b/src/condor_submit.V6/submit.cpp
@@ -7462,7 +7462,7 @@ SetVMParams()
 	tmp_ptr = condor_param(VM_MACAddr, ATTR_JOB_VM_MACADDR);
 	if(tmp_ptr)
 	  {
-	    buffer.sprintf("%s = %d", ATTR_JOB_VM_MACADDR, tmp_ptr);
+	    buffer.sprintf("%s = \"%s\"", ATTR_JOB_VM_MACADDR, tmp_ptr);
 	    InsertJobExpr(buffer, false);
 	  }
 
