diff --git a/src/condor_starter.V6.1/vm_gahp_server.cpp b/src/condor_starter.V6.1/vm_gahp_server.cpp
index f479d67..5863a63 100644
--- a/src/condor_starter.V6.1/vm_gahp_server.cpp
+++ b/src/condor_starter.V6.1/vm_gahp_server.cpp
@@ -86,7 +86,7 @@ VMGahpServer::VMGahpServer(const char *vmgahpserver,
 	m_include_gahp_log = false;
 #endif
 
-	m_send_all_classad = param_boolean("VM_GAHP_SEND_ALL_CLASSAD", false);
+	m_send_all_classad = param_boolean("VM_GAHP_SEND_ALL_CLASSAD", true);
 
 	if( m_send_all_classad ) {
 		dprintf( D_FULLDEBUG, "Will send the entire job ClassAd to vmgahp\n");
