diff --git a/src/condor_io/shared_port_client.cpp b/src/condor_io/shared_port_client.cpp
index d19d209..5d3af95 100644
--- a/src/condor_io/shared_port_client.cpp
+++ b/src/condor_io/shared_port_client.cpp
@@ -407,7 +407,7 @@ SharedPortState::Handle(Stream *s)
 
 	// If we are done, clean up and dellocate
 	if (result == DONE || result == FAILED) {
-		if (s && (m_state != RECV_RESP)) {
+		if (m_state != RECV_RESP || !m_non_blocking) {
 			delete s;
 		}
 		delete this;
