diff --git a/src/condor_utils/file_lock.cpp b/src/condor_utils/file_lock.cpp
index 15e09ff..09a299b 100644
--- a/src/condor_utils/file_lock.cpp
+++ b/src/condor_utils/file_lock.cpp
@@ -705,10 +705,9 @@ FileLock::CreateHashName(const char *orig, bool useDefault)
 		c = temp_filename[i];
 		hash = c + (hash << 6) + (hash << 16) - hash;
 	}
-	char hashVal[256] = {0};
+	hash = hash%10000;
+	char hashVal[7] = {0};
 	sprintf(hashVal, "%u", hash);
-	while (strlen(hashVal) < 5)
-		sprintf(hashVal+strlen(hashVal), "%s", hashVal);
 
 	int len = strlen(path) + strlen(hashVal) + 20;
 	
@@ -721,12 +720,11 @@ FileLock::CreateHashName(const char *orig, bool useDefault)
 		sprintf(dest, "%s", path  );
 	delete []temp_filename; 
 	delete []path;
-	for (int i = 0 ; i < 4; i+=2 ) {
-		snprintf(dest+strlen(dest), 3, "%s", hashVal+i);
-		snprintf(dest+strlen(dest), 2, "%c", DIR_DELIM_CHAR);
-	}
-	 
 	
-	sprintf(dest+strlen(dest), "%s.lockc", hashVal+4);
+#if !defined(WIN32)
+	sprintf(dest+strlen(dest), "%s%c%i%c", hashVal, DIR_DELIM_CHAR, geteuid(), DIR_DELIM_CHAR);
+#endif
+	
+	sprintf(dest+strlen(dest), "%s.lockc", hashVal);
 	return dest;
 }
