diff --git a/src/condor_negotiator.V6/matchmaker.cpp b/src/condor_negotiator.V6/matchmaker.cpp
index 8516c91..914328c 100644
--- a/src/condor_negotiator.V6/matchmaker.cpp
+++ b/src/condor_negotiator.V6/matchmaker.cpp
@@ -4457,12 +4457,17 @@ insertClaimRequirements(ClassAd &request, ClassAd *offer)
     }
   }
 
-  { // optional concurrency limits constraint
+  { // concurrency limits constraint
     MyString limits;
     if (request.LookupString(ATTR_CONCURRENCY_LIMITS, limits)) {
+      // request exactly the same limits, if the current one has any
       limits.lower_case();
       std::string req=std::string("(toLower(TARGET.")+ATTR_CONCURRENCY_LIMITS+")=?=\""+limits+"\")";
       req_list.push_back(req);
+    } else {
+      // or prevent jobs with any limits to match
+      std::string req=std::string("TARGET.")+ATTR_CONCURRENCY_LIMITS+"=?=UNDEFINED";
+      req_list.push_back(req);
     }
   }
   
