diff --git a/src/condor_includes/condor_auth_x509.h b/src/condor_includes/condor_auth_x509.h
index e137dcc..a13136a 100644
--- a/src/condor_includes/condor_auth_x509.h
+++ b/src/condor_includes/condor_auth_x509.h
@@ -144,6 +144,7 @@ class Condor_Auth_X509 : public Condor_Auth_Base {
     typedef HashTable<MyString, MyString> Grid_Map_t;
     static Grid_Map_t * GridMap;
 #endif
+	static bool m_globusActivated;
 };
 
 #endif
diff --git a/src/condor_io/condor_auth_x509.cpp b/src/condor_io/condor_auth_x509.cpp
index fc68941..c7d2f0e 100644
--- a/src/condor_io/condor_auth_x509.cpp
+++ b/src/condor_io/condor_auth_x509.cpp
@@ -48,6 +48,8 @@ StringList * getDaemonList(ReliSock * sock);
 HashTable<MyString, MyString> * Condor_Auth_X509::GridMap = 0;
 #endif
 
+bool Condor_Auth_X509::m_globusActivated = false;
+
 //----------------------------------------------------------------------
 // Implementation
 //----------------------------------------------------------------------
@@ -62,6 +64,11 @@ Condor_Auth_X509 :: Condor_Auth_X509(ReliSock * sock)
 #ifdef WIN32
 	ParseMapFile();
 #endif
+	if ( !m_globusActivated ) {
+		globus_module_activate( GLOBUS_GSI_GSSAPI_MODULE );
+		globus_module_activate( GLOBUS_GSI_GSS_ASSIST_MODULE );
+		m_globusActivated = true;
+	}
 }
 
 Condor_Auth_X509 ::  ~Condor_Auth_X509()
