diff --git a/src/condor_utils/CMakeLists.txt b/src/condor_utils/CMakeLists.txt
index d5beaf6..1eb7838 100644
--- a/src/condor_utils/CMakeLists.txt
+++ b/src/condor_utils/CMakeLists.txt
@@ -72,9 +72,9 @@ endif()
 ##################################################
 # condorapi & tests
 
-condor_selective_glob("my_username.*;condor_event.*;file_sql.*;misc_utils.*;user_log_header.*;write_user_log*;read_user_log*;iso_dates.*;file_lock.*;format_time.*;utc_time.*;stat_wrapper*;log_rotate.*;dprintf*;sig_install.*;basename.*;mkargv.*;except.*;strupr.*;lock_file.*;rotate_file.*;strcasestr.*;strnewp.*;condor_environ.*;setsyscalls.*;passwd_cache.*;uids.c*;chomp.*;subsystem_info.*;my_subsystem.*;distribution.*;my_distribution.*;get_random_num.*;libcondorapi_stubs.*;seteuid.*;setegid.*;condor_open.*;classad_merge.*;condor_attributes.*;simple_arg.*;compat_classad.*;compat_classad_util.*;classad_oldnew.*;condor_snutils.*;stringSpace.*;string_list.*;stl_string_utils.*;MyString.*;condor_xml_classads.*;directory*;param_functions.*;${SAFE_OPEN_SRC}" ApiSrcs)
+condor_selective_glob("my_username.*;condor_event.*;file_sql.*;misc_utils.*;user_log_header.*;write_user_log*;read_user_log*;iso_dates.*;file_lock.*;format_time.*;utc_time.*;stat_wrapper*;log_rotate.*;dprintf*;sig_install.*;basename.*;mkargv.*;except.*;strupr.*;lock_file.*;rotate_file.*;strcasestr.*;strnewp.*;condor_environ.*;setsyscalls.*;passwd_cache.*;uids.c*;chomp.*;subsystem_info.*;my_subsystem.*;distribution.*;my_distribution.*;get_random_num.*;libcondorapi_stubs.*;seteuid.*;setegid.*;condor_open.*;classad_merge.*;condor_attributes.*;simple_arg.*;compat_classad.*;compat_classad_util.*;classad_oldnew.*;condor_snutils.*;stringSpace.*;string_list.*;stl_string_utils.*;MyString.*;condor_xml_classads.*;directory*;param_functions.*;filename_tools_cpp.*;filename_tools.*;stat_info.*;${SAFE_OPEN_SRC}" ApiSrcs)
 if(WINDOWS)
-    condor_selective_glob("directory.WINDOWS.*;directory_util.*;dynuser.WINDOWS.*;filename_tools.*;lock_file.WINDOWS.*;lsa_mgr.*;my_dynuser.*;ntsysinfo.WINDOWS.*;posix.WINDOWS.*;stat.WINDOWS.*;store_cred.*;token_cache.WINDOWS.*;truncate.WINDOWS.*" ApiSrcs)
+    condor_selective_glob("directory.WINDOWS.*;directory_util.*;dynuser.WINDOWS.*;lock_file.WINDOWS.*;lsa_mgr.*;my_dynuser.*;ntsysinfo.WINDOWS.*;posix.WINDOWS.*;stat.WINDOWS.*;store_cred.*;token_cache.WINDOWS.*;truncate.WINDOWS.*" ApiSrcs)
     set_property( TARGET utils_genparams PROPERTY FOLDER "libraries" )
 endif()
 
@@ -84,9 +84,13 @@ add_dependencies( condorapi condor_utils )
 if (HAVE_LIBDL)
 	set(APILIBDEP dl)
 endif()
-
 target_link_libraries(condorapi ${CLASSADS_FOUND};${APILIBDEP})
 
+condor_shared_lib( condorapi_shared "${ApiSrcs}" )
+add_dependencies( condorapi_shared condor_utils )
+target_link_libraries(condorapi_shared ${CLASSADS_FOUND};${APILIBDEP})
+set_target_properties( condorapi_shared PROPERTIES OUTPUT_NAME condorapi )
+
 if (WANT_FULL_DEPLOYMENT)
 	install (FILES MyString.h
 			 condor_event.h
@@ -102,6 +106,7 @@ if (WANT_FULL_DEPLOYMENT)
 			DESTINATION ${C_INCLUDE})
 
     install ( TARGETS condorapi DESTINATION ${C_LIB} )
+    install ( TARGETS condorapi_shared DESTINATION ${C_LIB} )
 endif(WANT_FULL_DEPLOYMENT)
 
 add_definitions(-DENABLE_STATE_DUMP)
diff --git a/src/condor_utils/condor_event.h b/src/condor_utils/condor_event.h
index 221617a..de30b59 100644
--- a/src/condor_utils/condor_event.h
+++ b/src/condor_utils/condor_event.h
@@ -25,7 +25,7 @@
 
 #include <stdio.h>              /* for FILE type */
 #if !defined(WIN32) 
-#include <sys/time.h>
+#include <time.h>
 #include <sys/resource.h>       /* for struct rusage */
 #endif
 
diff --git a/src/condor_utils/libcondorapi_stubs.cpp b/src/condor_utils/libcondorapi_stubs.cpp
index 9c5eee7..7fa8660 100644
--- a/src/condor_utils/libcondorapi_stubs.cpp
+++ b/src/condor_utils/libcondorapi_stubs.cpp
@@ -396,5 +396,10 @@ int pthread_sigmask(int how, const sigset_t *newmask, sigset_t *oldmask)
 }
 #endif
 
+int my_spawnl( const char* cmd, ... )
+{ return not_impl(); }
 
 END_C_DECLS
+
+void statusString( int status, MyString &str )
+{ not_impl(); }
diff --git a/src/condor_utils/write_user_log.h b/src/condor_utils/write_user_log.h
index 63e2dac..4faedcf 100644
--- a/src/condor_utils/write_user_log.h
+++ b/src/condor_utils/write_user_log.h
@@ -31,6 +31,7 @@
 #ifdef HAVE_SYS_TYPES_H
 #  include <sys/types.h>
 #endif
+#include <unistd.h>
 
 /* Predeclare some classes */
 class MyString;
