diff --git a/src/condor_starter.V6.1/jic_shadow.cpp b/src/condor_starter.V6.1/jic_shadow.cpp
index 533e324..8086d8f 100644
--- a/src/condor_starter.V6.1/jic_shadow.cpp
+++ b/src/condor_starter.V6.1/jic_shadow.cpp
@@ -1394,11 +1394,13 @@ JICShadow::initWithFileTransfer()
 		job_ad->LookupString( ATTR_JOB_OUTPUT, stdout_name );
 		job_ad->LookupString( ATTR_JOB_ERROR, stderr_name );
 		if ( job_ad->LookupBool( ATTR_STREAM_OUTPUT, stream ) && !stream &&
-			 !nullFile( stdout_name.c_str() ) ) {
+			 !nullFile( stdout_name.c_str() ) &&
+			 condor_basename(stdout_name.c_str()) != stdout_name.c_str() ) {
 			job_ad->Assign( ATTR_JOB_OUTPUT, StdoutRemapName );
 		}
 		if ( job_ad->LookupBool( ATTR_STREAM_ERROR, stream ) && !stream &&
-			 !nullFile( stderr_name.c_str() ) ) {
+			 !nullFile( stderr_name.c_str() ) &&
+			 condor_basename(stderr_name.c_str()) != stderr_name.c_str() ) {
 			if ( stdout_name == stderr_name ) {
 				job_ad->Assign( ATTR_JOB_ERROR, StdoutRemapName );
 			} else {
