diff --git a/src/condor_vm-gahp/xen_type.linux.cpp b/src/condor_vm-gahp/xen_type.linux.cpp
index 1aa9299..63bf091 100644
--- a/src/condor_vm-gahp/xen_type.linux.cpp
+++ b/src/condor_vm-gahp/xen_type.linux.cpp
@@ -651,7 +651,13 @@ VirshType::Status()
 				case (VIR_ERR_NO_DOMAIN):
 					// The VM isn't there anymore, so signal shutdown
 					vmprintf(D_FULLDEBUG, "Couldn't find domain %s, assuming it was shutdown\n", m_vm_name.Value());
-					m_self_shutdown = true;
+					if(getVMStatus() == VM_RUNNING) {
+						m_self_shutdown = true;
+					}
+					if(getVMStatus() != VM_STOPPED) {
+						setVMStatus(VM_STOPPED);
+						m_stop_time.getTime();
+					}
 					m_result_msg += "Stopped";
 					return true;
 				break;
@@ -661,7 +667,13 @@ VirshType::Status()
 					if ( NULL == ( dom = virDomainLookupByName(m_libvirt_connection, m_vm_name.Value() ) ) )
 					{
 						vmprintf(D_ALWAYS, "could not reconnect to libvirt... marking vm as stopped (should exit)\n");
-						m_self_shutdown = true;
+						if(getVMStatus() == VM_RUNNING) {
+							m_self_shutdown = true;
+						}
+						if(getVMStatus() != VM_STOPPED) {
+							setVMStatus(VM_STOPPED);
+							m_stop_time.getTime();
+						}
 						m_result_msg += "Stopped";
 						return true;
 					}
