diff -rNup condor-7.4.4.orig/src/condor_dagman/dagman_main.cpp condor-7.4.4/src/condor_dagman/dagman_main.cpp
--- condor-7.4.4.orig/src/condor_dagman/dagman_main.cpp	2010-10-11 17:37:00.000000000 -0400
+++ condor-7.4.4/src/condor_dagman/dagman_main.cpp	2011-04-18 17:19:34.000000000 -0400
@@ -886,9 +887,29 @@ int main_init (int argc, char ** const a
     	debug_printf( DEBUG_VERBOSE, "Parsing %s ...\n", dagFile );
 
     	if( !parse( dagman.dag, dagFile, dagman.useDagDir ) ) {
-				// Note: debug_error calls DC_Exit().
-        	debug_error( 1, DEBUG_QUIET, "Failed to parse %s\n",
-					 	dagFile );
+	       debug_printf( DEBUG_QUIET, "Failed to parse %s\n",
+			dagFile );
+		// if duplicate DAGMan is no longer alive	
+	       if (util_check_lock_file(lockFileName) == 0) {
+			ArgList args;
+			MyString constraint;
+
+			args.Clear();
+			args.AppendArg( dagman.condorRmExe);
+			args.AppendArg( "-const" );
+
+			constraint.sprintf( "%s == %d", ATTR_DAGMAN_JOB_ID,
+					dagman.DAGManJobId._cluster );
+			args.AppendArg( constraint.Value() );
+			if ( util_popen( args ) ) {
+				debug_printf( DEBUG_NORMAL, "Error removing DAGMan jobs\n");
+			}
+					
+		} 
+		unlink( lockFileName );
+		dagman.CleanUp();
+		DC_Exit( EXIT_ERROR );	
+	
     	}
 	}
 
