diff --git a/src/condor_collector.V6/collector.cpp b/src/condor_collector.V6/collector.cpp
index 433e9b0..08457e1 100644
--- a/src/condor_collector.V6/collector.cpp
+++ b/src/condor_collector.V6/collector.cpp
@@ -1465,11 +1465,17 @@ void CollectorDaemon::init_classad(int interval)
         free( tmp );
     }
 
+    MyString id;
     if( CollectorName ) {
-            ad->Assign( ATTR_NAME, CollectorName );
+            if( strchr( CollectorName, '@' ) ) {
+               id.sprintf( "%s", CollectorName );
+            } else {
+               id.sprintf( "%s@%s", CollectorName, my_full_hostname() );
+            }
     } else {
-            ad->Assign( ATTR_NAME, my_full_hostname() );
+            id.sprintf( "%s", my_full_hostname() );
     }
+    ad->Assign( ATTR_NAME, id.Value() );
 
     ad->Assign( ATTR_COLLECTOR_IP_ADDR, global_dc_sinful() );
 
