--- condor_submit.sh.old	2013-03-29 11:20:23.998099722 -0500
+++ condor_submit.sh	2013-03-29 12:27:56.804856554 -0500
@@ -53,7 +53,7 @@
 # Name of local requirements file: currently unused
 req_file=""
 
-while getopts "a:i:o:de:j:n:v:V:c:w:x:u:q:r:s:T:I:O:R:C:" arg 
+while getopts "a:i:o:de:j:n:v:V:c:w:x:u:q:r:s:T:I:O:R:C:m:" arg 
 do
     case "$arg" in
     a) xtra_args="$OPTARG" ;;
@@ -77,6 +77,7 @@
     O) outputflstring="$OPTARG" ;;
     R) remaps="$OPTARG" ;;
     C) req_file="$OPTARG" ;;
+    m) req_mem="$OPTARG" ;;
     -) break ;;
     ?) echo $usage_string
        exit 1 ;;
@@ -239,6 +240,12 @@
   echo -e $xtra_args >> $submit_file
 fi
 
+if [ "x$req_mem" != "x"]
+then
+  echo "request_memory = $req_mem" >> $submit_file
+fi
+
+
 cat >> $submit_file << EOF
 # We insist on new style quoting in Condor
 arguments = $arguments
