diff --git a/src/condor_c++_util/uids.C b/src/condor_c++_util/uids.C
index 8c25066..ca682f8 100644
--- a/src/condor_c++_util/uids.C
+++ b/src/condor_c++_util/uids.C
@@ -1131,6 +1131,11 @@ _set_priv(priv_state s, char file[], int line, int dologging)
 				"warning: attempted switch out of PRIV_USER_FINAL\n");
 		return PRIV_USER_FINAL;
 	}
+	if (CurrentPrivState == PRIV_CONDOR_FINAL) {
+		dprintf(D_ALWAYS,
+				"warning: attempted switch out of PRIV_CONDOR_FINAL\n");
+		return PRIV_CONDOR_FINAL;
+	}
 	CurrentPrivState = s;
 
 	if (can_switch_ids()) {
diff --git a/src/condor_daemon_core.V6/daemon_core.C b/src/condor_daemon_core.V6/daemon_core.C
index 9ea6d02..1fb6443 100644
--- a/src/condor_daemon_core.V6/daemon_core.C
+++ b/src/condor_daemon_core.V6/daemon_core.C
@@ -6562,13 +6562,16 @@ int DaemonCore::Create_Process(
 
 		// We have to do some checks on the executable name and the
 		// cwd before we fork.  We want to do these in priv state
-		// specified, but in the user priv if PRIV_USER_FINAL specified.
+		// specified, but in the user priv if PRIV_USER_FINAL specified
+		// or in the condor priv if PRIV_CONDOR_FINAL is specified.
 		// Don't do anything in PRIV_UNKNOWN case.
 
 	priv_state current_priv;
 	if ( priv != PRIV_UNKNOWN ) {
 		if ( priv == PRIV_USER_FINAL ) {
 			current_priv = set_user_priv();
+		} else if ( priv == PRIV_CONDOR_FINAL ) {
+			current_priv = set_condor_priv();
 		} else {
 			current_priv = set_priv( priv );
 		}
