--- pbs_submit.sh.old	2013-03-29 12:30:51.870854864 -0500
+++ pbs_submit.sh	2013-03-29 11:22:48.579852982 -0500
@@ -113,6 +113,14 @@
 
 bls_set_up_local_and_extra_args
 
+if [ "x$bls_opt_req_mem" != "x" ]
+then
+  # Different schedulers require different memory checks
+  echo "#PBS -l mem=$bls_opt_req_mem" >> $bls_tmp_file
+  echo "#PBS -l pmem=$bls_opt_req_mem" >> $bls_tmp_file
+  echo "#PBS -l pvmem=$bls_opt_req_mem" >> $bls_tmp_file
+fi
+
 # Write PBS directives according to command line options
 # handle queue overriding
 [ -z "$bls_opt_queue" ] || grep -q "^#PBS -q" $bls_tmp_file || echo "#PBS -q $bls_opt_queue" >> $bls_tmp_file
