diff --git a/src/condor_c++_util/param_info.in b/src/condor_c++_util/param_info.in
index 40366f8..9096895 100644
--- a/src/condor_c++_util/param_info.in
+++ b/src/condor_c++_util/param_info.in
@@ -802,38 +802,6 @@ url=http://cs.wisc.edu/condor/manual/v7.3/3_3Configuration.html#SECTION
 tags=vm-gahp,xen_type
 id=8
 
-[XEN_DEFAULT_KERNEL]
-aliases=
-default=
-version=0.0.0
-range=.*
-state=default
-type=string
-is_macro=false
-reconfig=true
-customization=seldom
-friendly_name=Xen Default Kernel
-usage=
-url=http://cs.wisc.edu/condor/manual/v7.3/3_3Configuration.html#SECTION
-tags=vm-gahp,xen_type
-id=9
-
-[XEN_DEFAULT_INITRD]
-aliases=
-default=
-version=0.0.0
-range=.*
-state=default
-type=string
-is_macro=false
-reconfig=true
-customization=seldom
-friendly_name=Xen Default Initrd
-usage=
-url=http://cs.wisc.edu/condor/manual/v7.3/3_3Configuration.html#SECTION
-tags=vm-gahp,xen_type
-id=10
-
 [XEN_BOOTLOADER]
 aliases=
 default=
diff --git a/src/condor_examples/condor_config.generic b/src/condor_examples/condor_config.generic
index dc20dcb..6f1bb21 100644
--- a/src/condor_examples/condor_config.generic
+++ b/src/condor_examples/condor_config.generic
@@ -2344,22 +2344,6 @@ LIBVIRT_XML_SCRIPT = $(LIBEXEC)/libvirt_simple_script.awk
 ##  VM Universe Parameters Specific to Xen
 ##--------------------------------------------------------------------
 
-##  Where is the default kernel image for Xen domainU? (Required)
-##
-##  This kernel image will be used when a job submission user doesn't
-##  define a kernel image explicitly. 
-##  If a job submission user defines its own kernel image in the submit
-##  description file, it will be used instead of this default kernel.
-##  (e.g.) XEN_DEFAULT_KERNEL = /boot/vmlinuz-2.6.16-xen
-#XEN_DEFAULT_KERNEL = /full/path/to/<kernel-image-file>
-
-##  Where is the default initrd image for Xen domainU? (Optional)
-##
-##  The complete path and image file name for the initrd image, if
-##  needed for the default kernel image. 
-##  (e.g.) XEN_DEFAULT_INITRD = /boot/initrd-2.6.16-xen0.img
-#XEN_DEFAULT_INITRD = /full/path/to/<initrd-image-file>
-
 ##  Where is bootloader for Xen domainU? (Required)
 ##
 ##  The bootloader will be used in the case that a kernel image includes
diff --git a/src/condor_includes/condor_vm_universe_types.h b/src/condor_includes/condor_vm_universe_types.h
index 000ed5a..67befa2 100644
--- a/src/condor_includes/condor_vm_universe_types.h
+++ b/src/condor_includes/condor_vm_universe_types.h
@@ -47,7 +47,6 @@
 #define VMGAHP_STATUS_COMMAND_CPUTIME	"CPUTIME"
 
 /* Parameters for Xen kernel */
-#define XEN_KERNEL_ANY			"any"
 #define XEN_KERNEL_INCLUDED		"included"
 #define XEN_KERNEL_HW_VT		"vmx"
 
@@ -60,7 +59,6 @@
 #define VMPARAM_XEN_CDROM_DEVICE	"VMPARAM_Xen_CDROM_Device"
 #define VMPARAM_XEN_TRANSFER_FILES	"VMPARAM_Xen_Transfer_Files"
 #define VMPARAM_XEN_BOOTLOADER		"VMPARAM_Xen_Bootloader"
-#define VMPARAM_XEN_KERNEL_IMAGE	"VMPARAM_Xen_Kernel_Image"
 
 /* ClassAd Attributes for KVM */
 #define VMPARAM_KVM_DISK			"VMPARAM_Kvm_Disk"
diff --git a/src/condor_submit.V6/submit.cpp b/src/condor_submit.V6/submit.cpp
index 8e172a5..a3c3759 100644
--- a/src/condor_submit.V6/submit.cpp
+++ b/src/condor_submit.V6/submit.cpp
@@ -7734,19 +7734,14 @@ SetVMParams()
 						"Please specify 'xen_kernel' for the xen virtual machine "
 						"in your submit description file.\n"
 						"xen_kernel must be one of "
-						"\"%s\", \"%s\", \"%s\", <file-name>.\n", 
-						XEN_KERNEL_ANY, XEN_KERNEL_INCLUDED, XEN_KERNEL_HW_VT);
+						"\"%s\", \"%s\", <file-name>.\n", 
+						XEN_KERNEL_INCLUDED, XEN_KERNEL_HW_VT);
 				DoCleanup(0,0,NULL);
 				exit(1);
 			}else {
 				MyString fixedname = delete_quotation_marks(xen_kernel);
 
-				if( stricmp(fixedname.Value(), XEN_KERNEL_ANY ) == 0 ) {
-					// Condor will use a default kernel defined 
-					// in a vmgahp config file on an execute machine
-					real_xen_kernel_file = false;
-					need_xen_root_device = true;
-				}else if ( stricmp(fixedname.Value(), XEN_KERNEL_INCLUDED) == 0) {
+				if ( stricmp(fixedname.Value(), XEN_KERNEL_INCLUDED) == 0) {
 					// kernel image is included in a disk image file
 					// so we will use bootloader(pygrub etc.) defined 
 					// in a vmgahp config file on an excute machine 
diff --git a/src/condor_vm-gahp/libvirt_simple_script.awk b/src/condor_vm-gahp/libvirt_simple_script.awk
index b7b29f7..628c698 100644
--- a/src/condor_vm-gahp/libvirt_simple_script.awk
+++ b/src/condor_vm-gahp/libvirt_simple_script.awk
@@ -44,14 +44,7 @@ END {
 	}
 	else
 	{
-	    if(index(attrs["VMPARAM_Xen_Kernel"],"any") != 0)
-	    {
-		kernel = attrs["VMPARAM_Xen_Kernel_Image"]
-	    }
-	    else
-	    {
-		kernel = attrs["VMPARAM_Xen_Kernel"]
-	    }
+	    kernel = attrs["VMPARAM_Xen_Kernel"]
 	    root = attrs["VMPARAM_Xen_Root"]
 	    initrd = attrs["VMPARAM_Xen_Initrd"]
             kern_params = attrs["VMPARAM_Xen_Kernel_Params"]
diff --git a/src/condor_vm-gahp/vmgahp_main.cpp b/src/condor_vm-gahp/vmgahp_main.cpp
index 789c58b..001e2ed 100644
--- a/src/condor_vm-gahp/vmgahp_main.cpp
+++ b/src/condor_vm-gahp/vmgahp_main.cpp
@@ -382,7 +382,7 @@ int main_init(int argc, char *argv[])
 	if( (strcasecmp(vmtype.Value(), CONDOR_VM_UNIVERSE_XEN) == 0) || (strcasecmp(vmtype.Value(), CONDOR_VM_UNIVERSE_KVM) == 0)) {
 		// Xen requires root priviledge 
 		if( !canSwitchUid() ) {
-			vmprintf(D_ALWAYS, "VMGahp server for Xen requires "
+			vmprintf(D_ALWAYS, "VMGahp server for Xen or KVM requires "
 					"root privilege\n");
 			DC_Exit(1);
 		}
diff --git a/src/condor_vm-gahp/xen_type.cpp b/src/condor_vm-gahp/xen_type.cpp
index ef700c8..5c7541c 100644
--- a/src/condor_vm-gahp/xen_type.cpp
+++ b/src/condor_vm-gahp/xen_type.cpp
@@ -341,10 +341,6 @@ bool VirshType::CreateVirshConfigFile(const char* filename)
   classad_string += " = \"";
   classad_string += m_xen_bootloader;
   classad_string += "\"\n";
-  classad_string += VMPARAM_XEN_KERNEL_IMAGE;
-  classad_string += " = \"";
-  classad_string += m_xen_kernel_file;
-  classad_string += "\"\n";
   if(classad_string.find(VMPARAM_XEN_INITRD) < 1)
     {
       classad_string += VMPARAM_XEN_INITRD;
@@ -1334,33 +1330,6 @@ XenType::checkXenParams(VMGahpConfig* config)
 	}
 	config->m_vm_script = fixedvalue;
 
-	// Read XEN_DEFAULT_KERNEL (required parameter)
-	config_value = param("XEN_DEFAULT_KERNEL");
-	if( !config_value ) {
-		vmprintf(D_ALWAYS, "\nERROR: You should define the default "
-				"kernel for Xen\n");
-		return false;
-	}
-	fixedvalue = delete_quotation_marks(config_value);
-	free(config_value);
-
-	// Can read default xen kernel ?
-	if( check_vm_read_access_file(fixedvalue.Value(), true) == false ) {
-		return false;
-	}
-	
-	// Read XEN_DEFAULT_INITRD (optional parameter)
-	config_value = param("XEN_DEFAULT_INITRD");
-	if( config_value ) {
-		fixedvalue = delete_quotation_marks(config_value);
-		free(config_value);
-
-		// Can read default xen ramdisk ?
-		if( check_vm_read_access_file(fixedvalue.Value(), true) == false ) {
-			return false;
-		}
-	}
-
 	// Read XEN_BOOTLOADER (required parameter)
 	config_value = param("XEN_BOOTLOADER");
 	if( !config_value ) {
@@ -1677,25 +1646,7 @@ bool XenType::CreateConfigFile()
 	}
 	m_xen_kernel_submit_param.trim();
 
-	if(strcasecmp(m_xen_kernel_submit_param.Value(), XEN_KERNEL_ANY) == 0) {
-		vmprintf(D_ALWAYS, "VMGahp will use default xen kernel\n");
-		config_value = param( "XEN_DEFAULT_KERNEL" );
-		if( !config_value ) {
-			vmprintf(D_ALWAYS, "Default xen kernel is not defined "
-					"in vmgahp config file\n");
-			m_result_msg = VMGAHP_ERR_CRITICAL;
-			return false;
-		}else {
-			m_xen_kernel_file = delete_quotation_marks(config_value);
-			free(config_value);
-
-			config_value = param( "XEN_DEFAULT_INITRD" );
-			if( config_value ) {
-				m_xen_initrd_file = delete_quotation_marks(config_value);
-				free(config_value);
-			}
-		}
-	}else if(strcasecmp(m_xen_kernel_submit_param.Value(), XEN_KERNEL_INCLUDED) == 0 )
+	if(strcasecmp(m_xen_kernel_submit_param.Value(), XEN_KERNEL_INCLUDED) == 0 )
 	{
 		//if (strcasecmp(vm_type.Value(), CONDOR_VM_UNIVERSE_XEN) == 0){
 			vmprintf(D_ALWAYS, "VMGahp will use xen bootloader\n");
