diff --git a/msconfig/WiX/xml/win.xsl.in b/msconfig/WiX/xml/win.xsl.in
index 6c37abf..7882b02 100755
--- a/msconfig/WiX/xml/win.xsl.in
+++ b/msconfig/WiX/xml/win.xsl.in
@@ -48,9 +48,10 @@
                  <RegistryValue Type="string" Name="HDFSMODE" Value="[HDFSMODE]" />
                  <RegistryValue Type="string" Name="HDFSPORT" Value="[HDFSPORT]" />
                  <RegistryValue Type="string" Name="HDFSWEBPORT" Value="[HDFSWEBPORT]" />
-             </RegistryKey>
-			 <RegistryKey Root="HKLM" Key="SYSTEM\CurrentControlSet\Services\condor"  Action="createAndRemoveOnUninstall">
+             </RegistryKey>
+             <RegistryKey Root="HKLM" Key="SYSTEM\CurrentControlSet\Services\condor"  Action="createAndRemoveOnUninstall">
                <RegistryValue Type="integer" Name="DelayedAutostart" Value="1" />
+               <RegistryValue Type="multiString" Name="DependOnService" Value="Dhcp" />
              </RegistryKey>
              <Environment Id="CondorBin" Action="set" Name="PATH" Part="last" Permanent="no" System="yes" Value="[INSTALLDIR]bin\"/>
            </Component>
