diff --git a/build/packaging/rpm/condor.spec.in b/build/packaging/rpm/condor.spec.in
index 41199bc..cfc2d9a 100644
--- a/build/packaging/rpm/condor.spec.in
+++ b/build/packaging/rpm/condor.spec.in
@@ -10,6 +10,7 @@ URL:		@CPACK_RPM_PACKAGE_URL@
 
 Requires(pre): @CPACK_RPM_SHADOW_PACKAGE@
 Requires(post):/sbin/chkconfig
+Requires(post): /usr/sbin/selinuxenabled, /usr/sbin/semanage, /usr/sbin/restorecon
 Requires(preun):/sbin/chkconfig
 Requires(preun):/sbin/service
 Requires(postun):/sbin/service
@@ -146,8 +147,7 @@ fi
 #Add condor service
 /sbin/chkconfig --add condor
 /sbin/ldconfig
-test -x /usr/sbin/selinuxenabled && /usr/sbin/selinuxenabled
-if [ $? = 0 ]; then
+if /usr/sbin/selinuxenabled; then
    semanage fcontext -a -t unconfined_execmem_exec_t $USR/sbin/condor_startd 2>&1| grep -v "already defined"
    restorecon  $USR/sbin/condor_startd
 fi
