diff --git a/src/classad/fnCall.cpp b/src/classad/fnCall.cpp
index ea2651d..3593321 100644
--- a/src/classad/fnCall.cpp
+++ b/src/classad/fnCall.cpp
@@ -1847,7 +1847,7 @@ unparse(const char*, const ArgumentList &argList, EvalState &state,
 	Value &result )
 {
 	
-	if( argList.size() != 1 ) {
+	if( argList.size() != 1 || argList[0]->GetKind() != ATTRREF_NODE ) {
 		result.SetErrorValue( );
 	}
 	else{
