commit 836a17b250506c753797b31a563a28d642c5d674
Author: Dan Bradley <dan@hep.wisc.edu>
Date:   Tue Oct 9 10:25:05 2012 -0500

    Fixed bug handling MODIFY_REQUEST_EXPR

diff --git a/src/condor_startd.V6/Resource.cpp b/src/condor_startd.V6/Resource.cpp
index 38b4db1..74295c7 100644
--- a/src/condor_startd.V6/Resource.cpp
+++ b/src/condor_startd.V6/Resource.cpp
@@ -2778,8 +2778,8 @@ Resource * initialize_resource(Resource * rip, ClassAd * req_classad, Claim* &le
 					// so we can try again.
 					dprintf(D_ALWAYS, 
 						"Job no longer matches partitionable slot after MODIFY_REQUEST_EXPR_ edits, retrying w/o edits\n");
-					if ( req_classad ) delete req_classad;	// delete modified ad
-					req_classad = unmodified_req_classad;	// put back original					
+					req_classad->CopyFrom(*unmodified_req_classad);	// put back original					
+					delete unmodified_req_classad;
 					unmodified_req_classad = NULL;
 				} else {
 					rip->dprintf(D_ALWAYS, 
