diff --git a/externals/bundles/blahp/1.16.1/CMakeLists.txt b/externals/bundles/blahp/1.16.1/CMakeLists.txt
index a4e7a2f..4c6e3c8 100644
--- a/externals/bundles/blahp/1.16.1/CMakeLists.txt
+++ b/externals/bundles/blahp/1.16.1/CMakeLists.txt
@@ -19,18 +19,19 @@
 # The source tarball for this external was downloaded from this URL:
 #   http://jra1mw.cvs.cern.ch/cgi-bin/jra1mw.cgi/org.glite.ce.blahp.tar.gz?view=tar&pathrev=glite-ce-blahp_R_1_16_1_0
 
-if (NOT PROPER AND LINUX)
-  option(WITH_BLAHP "Compiling with support for BLAHP" ON)
-endif(NOT PROPER AND LINUX)
+if ( LINUX )
+	option(WITH_BLAHP "Compiling with support for BLAHP" ON)
+endif ( LINUX )
 
 if ( WITH_BLAHP )
 
-	message(STATUS "TODO:TSTCLAIR BLAHP eval if linking correct popt w/o mundging Makefile.am")
-	
-	set(POPT_VER popt-1.14)
-	set(BLAHP_VER blahp-1.16.1)
+	if ( NOT PROPER )
+		message(STATUS "TODO:TSTCLAIR BLAHP eval if linking correct popt w/o mundging Makefile.am")
 
-	ExternalProject_Add(popt
+		set(POPT_VER popt-1.14)
+		set(BLAHP_VER blahp-1.16.1)
+
+		ExternalProject_Add(popt
 						PREFIX ${EXTERNAL_BUILD_PREFIX}/${POPT_VER}
 						#-- Download Step ----------
 						DOWNLOAD_COMMAND wget -N http://parrot.cs.wisc.edu/externals/${POPT_VER}.tar.gz ${CMD_TERM}
@@ -45,9 +46,9 @@ if ( WITH_BLAHP )
 						INSTALL_DIR ${EXTERNAL_STAGE}
 						INSTALL_COMMAND cd ${POPT_VER} && make install )
 
-	#dprint("we ingore the return param from bootstrap.. there are errors!")
+		#dprint("we ingore the return param from bootstrap.. there are errors!")
 
-	ExternalProject_Add(blahp
+		ExternalProject_Add(blahp
 						DEPENDS openssl globus popt
 					    PREFIX ${EXTERNAL_BUILD_PREFIX}/${BLAHP_VER}
 					    #-- Download Step ----------
@@ -77,17 +78,30 @@ if ( WITH_BLAHP )
 						#--install Step ----------
 						INSTALL_DIR ${EXTERNAL_STAGE}
 						INSTALL_COMMAND cd org.glite.ce.blahp && make install )
-	add_dependencies(blahp classads)	
 
-	message( STATUS "external configured (blahp)" )
-	set( HAVE_EXT_BLAHP ON PARENT_SCOPE )
-	append_var (CONDOR_EXTERNALS blahp)
+		### set the install
+		install ( DIRECTORY ${EXTERNAL_STAGE}/opt/glite DESTINATION ${C_LIB} )
+
+		add_dependencies(blahp classads)	
+
+		set(BLAHP_FOUND ON)
+
+	else ( NOT PROPER )
+
+		find_program( BLAHP_FOUND BLClient HINTS /usr/bin )
+
+	endif ( NOT PROPER )
+
+	if ( BLAHP_FOUND )
+
+		message( STATUS "external configured (blahp)" )
+		set( HAVE_EXT_BLAHP ON PARENT_SCOPE )
+		append_var (CONDOR_EXTERNALS blahp)
 
-	### set the install
-	install ( DIRECTORY ${EXTERNAL_STAGE}/opt/glite DESTINATION ${C_LIB} )
+	endif ( BLAHP_FOUND )
 
 else( WITH_BLAHP )
 
-	message ( STATUS "external skiped (blahp)" )
+	message ( STATUS "external skipped (blahp)" )
 
 endif( WITH_BLAHP )
