diff --git a/msconfig/condor_api.vcproj b/msconfig/condor_api.vcproj
index c6f0cda..e8f9499 100644
--- a/msconfig/condor_api.vcproj
+++ b/msconfig/condor_api.vcproj
@@ -97,15 +97,11 @@
 	</References>
 	<Files>
 		<File
-			RelativePath="$(OutDir)\ast.obj"
+			RelativePath="$(OutDir)\compat_classad.obj"
 			>
 		</File>
 		<File
-			RelativePath="$(OutDir)\astbase.obj"
-			>
-		</File>
-		<File
-			RelativePath="$(OutDir)\attrlist.obj"
+			RelativePath="$(OutDir)\compat_classad_util.obj"
 			>
 		</File>
 		<File
@@ -121,10 +117,6 @@
 			>
 		</File>
 		<File
-			RelativePath="$(OutDir)\classad.obj"
-			>
-		</File>
-		<File
 			RelativePath="$(OutDir)\classad_merge.obj"
 			>
 		</File>
@@ -165,14 +157,6 @@
 			>
 		</File>
 		<File
-			RelativePath="$(OutDir)\environment.obj"
-			>
-		</File>
-		<File
-			RelativePath="$(OutDir)\evaluateOperators.obj"
-			>
-		</File>
-		<File
 			RelativePath="$(OutDir)\except.obj"
 			>
 		</File>
@@ -225,14 +209,6 @@
 			>
 		</File>
 		<File
-			RelativePath="$(OutDir)\operators.obj"
-			>
-		</File>
-		<File
-			RelativePath="$(OutDir)\parser.obj"
-			>
-		</File>
-		<File
 			RelativePath="$(OutDir)\read_user_log.obj"
 			>
 		</File>
@@ -241,18 +217,10 @@
 			>
 		</File>
 		<File
-			RelativePath="$(OutDir)\registration.obj"
-			>
-		</File>
-		<File
 			RelativePath="$(OutDir)\rotate_file.obj"
 			>
 		</File>
 		<File
-			RelativePath="$(OutDir)\scanner.obj"
-			>
-		</File>
-		<File
 			RelativePath="$(OutDir)\setsyscalls.obj"
 			>
 		</File>
@@ -297,10 +265,6 @@
 			>
 		</File>
 		<File
-			RelativePath="$(OutDir)\value.obj"
-			>
-		</File>
-		<File
 			RelativePath="$(OutDir)\write_user_log.obj"
 			>
 		</File>
diff --git a/msconfig/condor_classad.vcproj b/msconfig/condor_classad.vcproj
index 233a217..2200119 100644
--- a/msconfig/condor_classad.vcproj
+++ b/msconfig/condor_classad.vcproj
@@ -95,31 +95,27 @@
 	</References>
 	<Files>
 		<File
-			RelativePath="..\src\classad.old\ast.cpp"
+			RelativePath="..\src\condor_c++_util\compat_classad.cpp"
 			>
 		</File>
 		<File
-			RelativePath="..\src\classad.old\astbase.cpp"
+			RelativePath="..\src\condor_c++_util\compat_classad_list.cpp"
 			>
 		</File>
 		<File
-			RelativePath="..\src\classad.old\attrlist.cpp"
+			RelativePath="..\src\condor_c++_util\compat_classad_util.cpp"
 			>
 		</File>
 		<File
-			RelativePath="..\src\classad.old\buildtable.cpp"
+			RelativePath="..\src\condor_c++_util\compat_classad.h"
 			>
 		</File>
 		<File
-			RelativePath="..\src\classad.old\classad.cpp"
+			RelativePath="..\src\condor_c++_util\compat_classad_list.h"
 			>
 		</File>
 		<File
-			RelativePath="..\src\classad.old\classad_list.cpp"
-			>
-		</File>
-		<File
-			RelativePath="..\src\classad.old\classad_util.cpp"
+			RelativePath="..\src\condor_c++_util\compat_classad_util.h"
 			>
 		</File>
 		<File
@@ -183,39 +179,7 @@
 			>
 		</File>
 		<File
-			RelativePath="..\src\classad.old\environment.cpp"
-			>
-		</File>
-		<File
-			RelativePath="..\src\classad.old\evaluateOperators.cpp"
-			>
-		</File>
-		<File
-			RelativePath="..\src\classad.old\operators.cpp"
-			>
-		</File>
-		<File
-			RelativePath="..\src\classad.old\operators.h"
-			>
-		</File>
-		<File
-			RelativePath="..\src\classad.old\parser.cpp"
-			>
-		</File>
-		<File
-			RelativePath="..\src\classad.old\registration.cpp"
-			>
-		</File>
-		<File
-			RelativePath="..\src\classad.old\scanner.cpp"
-			>
-		</File>
-		<File
-			RelativePath="..\src\classad.old\value.cpp"
-			>
-		</File>
-		<File
-			RelativePath="..\src\classad.old\value.h"
+			RelativePath="..\src\classad\classad\xmlSink.h"
 			>
 		</File>
 	</Files>
diff --git a/src/condor_c++_util/compat_classad.cpp b/src/condor_c++_util/compat_classad.cpp
index e959143..ccf8c0d 100644
--- a/src/condor_c++_util/compat_classad.cpp
+++ b/src/condor_c++_util/compat_classad.cpp
@@ -17,6 +17,7 @@
  *
  ***************************************************************/
 
+#include "condor_common.h"
 #include "compat_classad.h"
 
 #include "condor_classad.h"
diff --git a/src/condor_c++_util/compat_classad_list.cpp b/src/condor_c++_util/compat_classad_list.cpp
index 7b6575c..4831c2e 100644
--- a/src/condor_c++_util/compat_classad_list.cpp
+++ b/src/condor_c++_util/compat_classad_list.cpp
@@ -17,6 +17,7 @@
  *
  ***************************************************************/
 
+#include "condor_common.h"
 #include "compat_classad_list.h"
 #include "condor_xml_classads.h"
 
diff --git a/src/condor_c++_util/compat_classad_util.cpp b/src/condor_c++_util/compat_classad_util.cpp
index 8ba1814..6b653b8 100644
--- a/src/condor_c++_util/compat_classad_util.cpp
+++ b/src/condor_c++_util/compat_classad_util.cpp
@@ -17,6 +17,7 @@
  *
  ***************************************************************/
 
+#include "condor_common.h"
 #include "compat_classad_util.h"
 #include "classad_oldnew.h"
 
diff --git a/src/condor_includes/condor_sys_nt.h b/src/condor_includes/condor_sys_nt.h
index b66ee41..02b63ac 100644
--- a/src/condor_includes/condor_sys_nt.h
+++ b/src/condor_includes/condor_sys_nt.h
@@ -255,9 +255,4 @@ END_C_DECLS
 /* fix [f]stat on Windows */
 #include "stat.WINDOWS.h"
 
-// For now, always use old ClassAds
-#if !defined( WANT_OLD_CLASSADS )
-# define WANT_OLD_CLASSADS 1
-#endif
-
 #endif /* CONDOR_SYS_NT_H */
