diff -rNup condor-7.4.4.orig/src/condor_negotiator.V6/Accountant.cpp condor-7.4.4/src/condor_negotiator.V6/Accountant.cpp
--- condor-7.4.4.orig/src/condor_negotiator.V6/Accountant.cpp	2011-02-09 14:47:51.000000000 -0500
+++ condor-7.4.4/src/condor_negotiator.V6/Accountant.cpp	2011-02-21 15:11:56.000000000 -0500
@@ -992,11 +992,22 @@ AttrList* Accountant::ReportState(const 
   AcctLog->table.startIterations();
   while (AcctLog->table.iterate(HK,ResourceAd)) {
     HK.sprint(key);
+  
     if (strncmp(ResourceRecord.Value(),key,ResourceRecord.Length())) continue;
 
     if (ResourceAd->LookupString(RemoteUserAttr,tmp)==0) continue;
-    if (CustomerName!=MyString(tmp)) continue;
-
+    
+    // is the customer name a group.username@ string?
+    string GroupName (CustomerName.Value());
+    string::size_type cpos= GroupName.find_last_of('@');
+        
+    string ResGroupName = GetAssignedGroup(MyString(tmp))->name;
+     
+    if (cpos==string::npos){ //customer name is a group string
+        if (GroupName.compare(ResGroupName)!=0 ) continue;
+    } else { //customername is group.username@ string
+       if (CustomerName!=MyString(tmp)) continue;
+     
     ResourceName=key+ResourceRecord.Length();
     sprintf(tmp,"Name%d = \"%s\"",ResourceNum,ResourceName.Value());
     ad->Insert(tmp);
@@ -1004,7 +1015,7 @@ AttrList* Accountant::ReportState(const 
     if (ResourceAd->LookupInteger(StartTimeAttr,StartTime)==0) StartTime=0;
     sprintf(tmp,"StartTime%d = %d",ResourceNum,StartTime);
     ad->Insert(tmp);
-
+    }
     ResourceNum++;
 	if( NumResourcesRW ) {
 		float SlotWeight = 1.0;
