diff --git a/src/classad_analysis/analysis.cpp b/src/classad_analysis/analysis.cpp
index 9b8a0d4..b85b302 100644
--- a/src/classad_analysis/analysis.cpp
+++ b/src/classad_analysis/analysis.cpp
@@ -248,7 +248,6 @@ bool ClassAdAnalyzer::
 AnalyzeJobReqToBuffer( ClassAd *request, ClassAdList &offers, string &buffer )
 {
 	ResourceGroup     rg;
-    classad::ClassAd  *converted_classad;
     classad::ClassAd  *explicit_classad;
     bool              success;
 
@@ -259,13 +258,7 @@ AnalyzeJobReqToBuffer( ClassAd *request, ClassAdList &offers, string &buffer )
 		return true;
 	}
 
-    converted_classad = toNewClassAd( request );
-	if( converted_classad == NULL ) {
-		buffer += "Unable to process job ClassAd";
-		buffer += "\n";
-		return true;
-	}
-    explicit_classad  = AddExplicitTargets( converted_classad );
+	explicit_classad  = AddExplicitTargets( request );
 
     ensure_result_initialized(explicit_classad);
     
@@ -273,9 +266,7 @@ AnalyzeJobReqToBuffer( ClassAd *request, ClassAdList &offers, string &buffer )
     offers.Rewind();
     ClassAd *ad;
     while((ad = offers.Next())) {
-      classad::ClassAd *new_ad = toNewClassAd(ad);
-      result_add_machine(*new_ad);
-      delete new_ad;
+      result_add_machine(*ad);
 
       if (do_basic_analysis) {
 	BasicAnalyze(request, ad);
@@ -284,7 +275,6 @@ AnalyzeJobReqToBuffer( ClassAd *request, ClassAdList &offers, string &buffer )
 
 	success = AnalyzeJobReqToBuffer( explicit_classad, rg, buffer );
 
-    delete converted_classad;
     delete explicit_classad;
     return success;
 }
@@ -313,7 +303,6 @@ AnalyzeJobAttrsToBuffer( ClassAd *request, ClassAdList &offers,
 						 string &buffer )
 {
 	ResourceGroup     rg;
-    classad::ClassAd  *converted_classad;
     classad::ClassAd  *explicit_classad;
     bool              success;
 
@@ -324,17 +313,10 @@ AnalyzeJobAttrsToBuffer( ClassAd *request, ClassAdList &offers,
 		return true;
 	}
 
-    converted_classad = toNewClassAd( request );
-	if( converted_classad == NULL ) {
-		buffer += "Unable to process job ClassAd";
-		buffer += "\n";
-		return true;
-	}
-    explicit_classad  = AddExplicitTargets( converted_classad );
+	explicit_classad  = AddExplicitTargets( request );
     ensure_result_initialized(explicit_classad);
 	success = AnalyzeJobAttrsToBuffer( explicit_classad, rg, buffer );
 
-    delete converted_classad;
     delete explicit_classad;
     return success;
 }
@@ -938,14 +920,8 @@ MakeResourceGroup( ClassAdList &caList, ResourceGroup &rg )
         classad::ClassAd *converted_classad;
         classad::ClassAd *explicit_classad;
 
-        converted_classad = toNewClassAd(ad);
-		if( converted_classad == NULL ) {
-				// could not convert to new ClassAd (parse error)
-			return false;
-		}
-        explicit_classad  = AddExplicitTargets(converted_classad);
+		explicit_classad  = AddExplicitTargets(ad);
 		newList.Append(explicit_classad);
-        delete converted_classad;
 		ad = caList.Next( );
 	}
 	if( !rg.Init( newList ) ) {
diff --git a/src/classad_analysis/result.cpp b/src/classad_analysis/result.cpp
index 8e25271..f267068 100644
--- a/src/classad_analysis/result.cpp
+++ b/src/classad_analysis/result.cpp
@@ -82,9 +82,7 @@ namespace classad_analysis {
   }
 
   void result::add_explanation(matchmaking_failure_kind reason, ClassAd *a_machine) {
-    classad::ClassAd *new_machine_ad = toNewClassAd(a_machine);
-    my_explanation[reason].push_back(*new_machine_ad);
-    delete new_machine_ad;
+    my_explanation[reason].push_back(*a_machine);
   }
 
   void result::add_suggestion(suggestion sg) {
