From 766efa2bb3a156ab3729e76d7aee1d7bb5dba3d2 Mon Sep 17 00:00:00 2001
From: Peter MacKinnon <pmackinn@redhat.com>
Date: Tue, 16 Nov 2010 17:02:30 -0500
Subject: [PATCH] Remove static visibility from qmgmt_all_users_trusted

---
 src/condor_schedd.V6/qmgmt.cpp |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/src/condor_schedd.V6/qmgmt.cpp b/src/condor_schedd.V6/qmgmt.cpp
index 62422a0..de2bd14 100644
--- a/src/condor_schedd.V6/qmgmt.cpp
+++ b/src/condor_schedd.V6/qmgmt.cpp
@@ -130,7 +130,7 @@ static int flush_job_queue_log_delay = 0;
 static void HandleFlushJobQueueLogTimer();
 static void ScheduleJobQueueLogFlush();
 
-static bool qmgmt_all_users_trusted = false;
+bool qmgmt_all_users_trusted = false;
 static char	**super_users = NULL;
 static int	num_super_users = 0;
 static char *default_super_user =
-- 
1.7.2.3

