diff --git a/src/classad/classad.cpp b/src/classad/classad.cpp
index dc812de..6630e75 100644
--- a/src/classad/classad.cpp
+++ b/src/classad/classad.cpp
@@ -386,6 +386,7 @@ bool ClassAd::Insert( std::string& serialized_nvp)
   bool bRet = false;
   string name, szValue;
   size_t pos, npos, vpos;
+  size_t bpos = 0;
   
   // comes in as "name = value" "name= value" or "name =value"
   npos=pos=serialized_nvp.find("=");
@@ -393,11 +394,15 @@ bool ClassAd::Insert( std::string& serialized_nvp)
   // only try to process if the string is valid 
   if ( pos != string::npos  )
   {
-    if (serialized_nvp[pos-1] == ' ')
+    while (npos > 0 && serialized_nvp[npos-1] == ' ')
     {
       npos--;
     }
-    name = serialized_nvp.substr(0, npos);
+    while (bpos < npos && serialized_nvp[bpos] == ' ')
+    {
+      bpos++;
+    }
+    name = serialized_nvp.substr(bpos, npos - bpos);
 
     vpos=pos+1;
     if (serialized_nvp[vpos] == ' ')
