diff -rNup condor-7.6.0.orig/src/condor_submit.V6/submit.cpp condor-7.6.0/src/condor_submit.V6/submit.cpp
--- condor-7.6.0.orig/src/condor_submit.V6/submit.cpp	2011-12-20 09:00:55.000000000 -0500
+++ condor-7.6.0/src/condor_submit.V6/submit.cpp	2012-01-06 15:19:43.000000000 -0500
@@ -6278,14 +6278,15 @@ check_requirements( char const *orig, My
 	for( aptr = answer.Value(); *aptr; aptr++ ) {
 		if( strncasecmp(ATTR_MEMORY,aptr,5) == MATCH ) {
 				// We found "Memory", but we need to make sure that's
-				// not part of "VirtualMemory"...
-			if( aptr == answer.Value() ) {
-					// We're at the beginning, must be Memory, since
-					// there's nothing before it.
+				// not part of "VirtualMemory" or "requestmemory"				
+			if( aptr == answer.Value() || 
+					// Otherwise, it's safe to go back one position
+				*(aptr-1) == '.' ) {
+					// We're either the beginning with just Memory, or
+					// it was specified as target.memory
 				checks_mem = true;
-				break;
-			}
-				// Otherwise, it's safe to go back one position:
+				continue; 
+			}				
 			if( *(aptr-1) == 'l' || *(aptr-1) == 'L' ) {
 					// Must be VirtualMemory, keep searching...
 				continue;
@@ -6296,9 +6297,8 @@ check_requirements( char const *orig, My
 				checks_reqmem = true;
 				continue;
 			}	
-		
 			checks_mem = true;
-			break;
+			continue;
 		}
 	}
  
