diff --git a/src/condor_utils/proc_family_proxy.cpp b/src/condor_utils/proc_family_proxy.cpp
index 9b10421..1e20886 100644
--- a/src/condor_utils/proc_family_proxy.cpp
+++ b/src/condor_utils/proc_family_proxy.cpp
@@ -631,6 +631,10 @@ ProcFamilyProxy::recover_from_procd_error()
 	m_client = NULL;
 	int ntries = 5;
 
+	// If we were the one to bring up the ProcD, we'll be the one
+	// to restart it
+	bool try_restart = m_procd_pid != -1;
+
 	while (ntries > 0 && m_client == NULL) {
 	
 		ntries--;
@@ -638,15 +642,14 @@ ProcFamilyProxy::recover_from_procd_error()
 		// the ProcD has failed. we know this either because communication
 		// has failed or the ProcD's reaper has fired
 		//
-		if (m_procd_pid != -1) {
+		if (try_restart) {
 
-			// we were the one to bring up the ProcD, so we'll be the one
-			// to restart it
-			//
 			dprintf(D_ALWAYS, "attempting to restart the Procd\n");
 			m_procd_pid = -1;
 			if (!start_procd()) {
-				EXCEPT("unable to start the ProcD");
+				// We failed to restart the procd, don't bother trying
+				// to create a ProcFamilyClient
+				continue;
 			}
 		}
 		else {
