From 2a408fbb6341bf8c4b984e6d7462faca31dd22bc Mon Sep 17 00:00:00 2001
From: Ben Cotton <ben.cotton@cyclecomputing.com>
Date: Fri, 19 Jun 2015 13:02:28 -0400
Subject: [PATCH] Fix bind to allow Fedora builds to work


diff --git a/src/condor_contrib/aviary/src/common/EndpointPublisher.cpp b/src/condor_contrib/aviary/src/common/EndpointPublisher.cpp
index c96b539..4aef808 100644
--- a/src/condor_contrib/aviary/src/common/EndpointPublisher.cpp
+++ b/src/condor_contrib/aviary/src/common/EndpointPublisher.cpp
@@ -63,7 +63,7 @@ EndpointPublisher::init(const std::string& uri_suffix, bool for_ssl)
 
 	// grab an ephemeral port
 	ReliSock probe_sock;
-	if (-1 == probe_sock.bind(true,0)) {
+	if (-1 == probe_sock.bind(CP_IPV4, true, 0, false)) {
 		dprintf(D_ALWAYS,"EndpointPublisher is unable to obtain ANY ephemeral port from configured range! " \
 			"Check configured values of LOWPORT,HIGHPORT.\n");
 		return false;
-- 
2.3.5

