diff -rNup condor-7.9.1.orig/src/condor_utils/ipv6_hostname.cpp condor-7.9.1/src/condor_utils/ipv6_hostname.cpp
--- condor-7.9.1.orig/src/condor_utils/ipv6_hostname.cpp	2012-08-16 15:51:24.000000000 -0400
+++ condor-7.9.1/src/condor_utils/ipv6_hostname.cpp	2013-06-12 07:44:48.665564149 -0400
@@ -173,6 +173,9 @@ MyString get_fqdn_from_hostname(const My
 		}
 
 		hostent* h = gethostbyname(hostname.Value());
+		
+		if (h && h->h_name && strchr(h->h_name, '.'))
+			return h->h_name;
 		if (h && h->h_aliases && *h->h_aliases) {
 			for (char** alias = h->h_aliases; *alias; ++alias) {
 				if (strchr(*alias, '.'))
