diff --git a/src/condor_dagman/dagman_main.cpp b/src/condor_dagman/dagman_main.cpp
index 682b7d8..fcd5d1e 100644
--- a/src/condor_dagman/dagman_main.cpp
+++ b/src/condor_dagman/dagman_main.cpp
@@ -1289,6 +1289,9 @@ main_pre_dc_init( int, char*[] )
 {
 	DC_Skip_Auth_Init();
 	DC_Skip_Core_Init();
+#ifdef WIN32
+	_setmaxstdio(2048);
+#endif
 
 		// Convert the DAGMan log file name to an absolute path if it's
 		// not one already, so that we'll log things to the right file
