diff --git a/src/condor_master.V6/masterDaemon.cpp b/src/condor_master.V6/masterDaemon.cpp
index 2fb77f3..8f326f6 100644
--- a/src/condor_master.V6/masterDaemon.cpp
+++ b/src/condor_master.V6/masterDaemon.cpp
@@ -2331,8 +2331,7 @@ Daemons::StopDaemonsBeforeMasterStops()
 	std::map<std::string, class daemon*>::iterator iter;
 
 	for( iter = daemon_ptr.begin(); iter != daemon_ptr.end(); iter++ ) {
-		if( ( iter->second->pid || iter->second->IsHA() )
-			&& iter->second->runs_here )
+		if( iter->second->pid && iter->second->runs_here )
 		{
 			iter->second->Stop();
 			running++;
