commit da5eeaf2be00df36ff8b3252634587ee3d90fab8
Author: Dan Bradley <dan@hep.wisc.edu>
Date:   Fri Jul 13 17:12:07 2012 -0500

    Fix memory leak in hostname-preservation patch. #1605

diff --git a/src/condor_q.V6/queue.cpp b/src/condor_q.V6/queue.cpp
index 85ad9cf..a64d932 100644
--- a/src/condor_q.V6/queue.cpp
+++ b/src/condor_q.V6/queue.cpp
@@ -3055,11 +3055,12 @@ show_queue( const char* v1, const char* v2, const char* v3, const char* v4, cons
 		} else {
 				// fetch queue from schedd	
 			int fetchResult;
-			Daemon *schedd_daemon_ptr;
-			if (ad)
-				schedd_daemon_ptr = new Daemon(ad, DT_SCHEDD, pool ? pool->addr() : NULL );
-			else
-				schedd_daemon_ptr = new Daemon(DT_SCHEDD, scheddName, pool ? pool->addr() : NULL );
+			std::auto_ptr<Daemon> schedd_daemon_ptr;
+			if (ad) {
+				schedd_daemon_ptr.reset( new Daemon(ad, DT_SCHEDD, pool ? pool->addr() : NULL ) );
+			} else {
+				schedd_daemon_ptr.reset( new Daemon(DT_SCHEDD, scheddName, pool ? pool->addr() : NULL ) );
+			}
 			Daemon &schedd_daemon = *schedd_daemon_ptr;
 			if( (fetchResult = Q.fetchQueueFromDaemon(jobs, attrs, schedd_daemon, scheddVersion, &errstack) != Q_OK)) {
 			// The parse + fetch failed, print out why
