diff --git a/src/python-bindings/classad.cpp b/src/python-bindings/classad.cpp
index 3405546..378fa21 100644
--- a/src/python-bindings/classad.cpp
+++ b/src/python-bindings/classad.cpp
@@ -784,9 +784,17 @@ convert_python_to_exprtree(boost::python::object value)
         boost::python::object timestamp = py_import("calendar").attr("timegm")(value.attr("timetuple")());
         // Determine the UTC offset; timetuple above is in local time, but timegm assumes UTC.
         atime.secs = boost::python::extract<time_t>(timestamp);
-        int offset = classad::Literal::findOffset(atime.secs);
+        int offset;
+        boost::python::object tzinfo = value.attr("utcoffset")();
+        if ( tzinfo == boost::python::object() ) {
+            // None object means no tzinfo, assume local time zone.
+            offset = classad::Literal::findOffset(atime.secs);
+        } else {
+            boost::python::object tz_secs = tzinfo.attr("seconds");
+            offset = boost::python::extract<int>(tz_secs);
+        }
         atime.secs -= offset;
-        atime.offset = 0;
+        atime.offset = offset;
         classad::Value val; val.SetAbsoluteTimeValue(atime);
         return classad::Literal::MakeLiteral(val);
     }
