From 172ae03c5accb8cbfce0f0594cef6ef9f3bbae23 Mon Sep 17 00:00:00 2001
From: Matyas Selmeci <matyas@cs.wisc.edu>
Date: Wed, 8 Apr 2015 11:37:59 -0500
Subject: [PATCH] Have environment scripts set PYTHONPATH (GT #4993)

---
 src/condor_scripts/condor_configure | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/condor_scripts/condor_configure b/src/condor_scripts/condor_configure
index 3266d7f..8a93024 100755
--- a/src/condor_scripts/condor_configure
+++ b/src/condor_scripts/condor_configure
@@ -572,17 +572,20 @@ unless ( $opt_disable_env_scripts ) {
 	print SH "# The script should be sourced by /bin/sh or similar\n";
 	print SH "CONDOR_CONFIG=\"$config_file\"\n";
 	print SH "export CONDOR_CONFIG\n";
 	print SH "PATH=\"$release_dir/bin:$release_dir/sbin:\$PATH\"\n";
 	print SH "export PATH\n";
+	print SH "PYTHONPATH=\"$release_dir/lib/python:\$PYTHONPATH\"\n";
+	print SH "export PYTHONPATH\n";
 	close( SH );
 
 	my $env_csh = $opt_env_scripts_dir ."/condor.csh";
 	open( CSH, ">$env_csh" ) or die "Can't write to $env_csh";
 	print CSH "# The script should be sourced by /bin/csh or similar\n";
 	print CSH "setenv CONDOR_CONFIG \"$config_file\"\n";
 	print CSH "setenv PATH \"$release_dir/bin:$release_dir/sbin:\$PATH\"\n";
+	print CSH "setenv PYTHONPATH \"$release_dir/lib/python:\$PYTHONPATH\"\n";
 	close( CSH );
 
 	if ( $opt_bosco ) {
 		$env_sh = $opt_env_scripts_dir ."/bosco.sh";
 		$env_csh = $opt_env_scripts_dir ."/bosco.csh";
-- 
2.0.1

