diff --git a/src/condor_schedd.V6/qmgmt.cpp b/src/condor_schedd.V6/qmgmt.cpp
index 51b5d68..076d50a 100644
--- a/src/condor_schedd.V6/qmgmt.cpp
+++ b/src/condor_schedd.V6/qmgmt.cpp
@@ -2393,6 +2393,7 @@ CommitTransaction(SetAttributeFlags_t flags /* = 0 */)
 	// submit events into the EVENT_LOG here.
 	if ( !new_ad_keys.empty() ) {
 		int cluster_id;
+		int old_cluster_id = -10;
 		int proc_id;
 		ClassAd *procad;
 		ClassAd *clusterad;
@@ -2401,13 +2402,25 @@ CommitTransaction(SetAttributeFlags_t flags /* = 0 */)
 		WriteUserLog usr_log;
 	
 		std::list<std::string>::iterator it;
+		int counter = 0;
+		int ad_keys_size = new_ad_keys.size();
 		for( it = new_ad_keys.begin(); it != new_ad_keys.end(); it++ ) {
+			++counter;
 			char const *key = it->c_str();
 			StrToId(key,cluster_id,proc_id);
-
+			// do we want to fsync the userLog?
+			bool doFsync = false; 
 			if( proc_id == -1 ) {
 				continue; // skip over cluster ads
 			}
+			//we want to fsync per cluster
+			if (old_cluster_id == -10)
+				old_cluster_id = cluster_id;
+				
+			if ( (old_cluster_id != cluster_id) || (counter == ad_keys_size) ) {
+				doFsync = true;
+				old_cluster_id = cluster_id;
+			}
 
 			char cluster_key[PROC_ID_STR_BUFLEN];
 			IdToStr(cluster_id,-1,cluster_key);
@@ -2439,7 +2452,8 @@ CommitTransaction(SetAttributeFlags_t flags /* = 0 */)
 					procad->LookupString(ATTR_GLOBAL_JOB_ID, gjid);
 					procad->LookupBool(ATTR_ULOG_USE_XML, use_xml);
 					usr_log.setEnableUserLog(true);	
-					usr_log.setUseXML(use_xml);				
+					usr_log.setUseXML(use_xml);	
+					usr_log.setEnableFsync(doFsync);
 					usr_log.initialize(owner.c_str(), ntdomain.c_str(), simple_name.c_str(),
 						0, 0, 0, gjid.c_str());
 					
