Index: schedd.C
===================================================================
RCS file: /p/condor/repository/CONDOR_SRC/src/condor_schedd.V6/schedd.C,v
retrieving revision 1.318.2.6
diff -u -r1.318.2.6 schedd.C
--- schedd.C	28 Feb 2008 15:42:32 -0000	1.318.2.6
+++ schedd.C	6 Mar 2008 20:07:19 -0000
@@ -2354,30 +2354,32 @@
 		   server and see any files in the directory that were written
 		   on a different machine.
 		*/
-	MyString iwd;
 	MyString owner;
-	BOOLEAN is_nfs;
 
-	if ( job_ad->LookupString( ATTR_OWNER, owner ) &&
-		 job_ad->LookupString( ATTR_JOB_IWD, iwd ) &&
-		 fs_detect_nfs( iwd.Value(), &is_nfs ) == 0 && is_nfs ) {
+		// We're not Windows, so we don't need the NT Domain
+	if ( !job_ad->LookupString( ATTR_OWNER, owner ) ) {
+		dprintf( D_ALWAYS, "No owner for job %d.%d!?\n", cluster, proc );
+	} else if ( !init_user_ids( owner.Value(), NULL ) ) {
+		dprintf( D_ALWAYS, "init_user_ids() failed for user %s!\n",
+				 owner.Value() );
+	} else {
 
+		MyString iwd;
+		BOOLEAN is_nfs;
 		priv_state priv;
 
-		dprintf( D_FULLDEBUG, "(%d.%d) Forcing NFS sync of Iwd\n", cluster,
-				 proc );
+		priv = set_user_priv();
+
+		if ( job_ad->LookupString( ATTR_JOB_IWD, iwd ) &&
+			 fs_detect_nfs( iwd.Value(), &is_nfs ) == 0 && is_nfs ) {
+
+			dprintf( D_FULLDEBUG, "(%d.%d) Forcing NFS sync of Iwd\n", cluster,
+					 proc );
 
-			// We're not Windows, so we don't need the NT Domain
-		if ( !init_user_ids( owner.Value(), NULL ) ) {
-			dprintf( D_ALWAYS, "init_user_ids() failed for user %s!\n",
-					 owner.Value() );
-		} else {
 			int sync_fd;
 			MyString filename_template;
 			char *sync_filename;
 
-			priv = set_user_priv();
-
 			filename_template.sprintf( "%s/.condor_nfs_sync_XXXXXX",
 									   iwd.Value() );
 			sync_filename = strdup( filename_template.Value() );
@@ -2388,9 +2390,9 @@
 			}
 
 			free( sync_filename );
-
-			set_priv( priv );
 		}
+
+		set_priv( priv );
 	}
 #endif /* WIN32 */
 
