Index: src/condor_utils/dbms_utils.cpp
==================================================================
--- src/condor_utils/dbms_utils.cpp
+++ src/condor_utils/dbms_utils.cpp
@@ -78,11 +78,11 @@
 			 tm->tm_sec,
 			 my_timezone(tm->tm_isdst));	
 
 	switch(dt) {
 	case T_PGSQL:
-		rv.sprintf("'%s'", tsv);		
+		rv.formatstr("'%s'", tsv);		
 		break;
 	default:
 		break;
 	}
 
@@ -235,11 +235,11 @@
 	strncpy(port, ptr_colon+1, len);
 		// terminate the string properyly
 	port[strlen(ptr_colon+1)] = '\0';
 
 		// get the password from the .pgpass file
-	writePasswordFile.sprintf( "%s/.pgpass", spool);
+	writePasswordFile.formatstr( "%s/.pgpass", spool);
 
 	writePassword = getWritePassword(writePasswordFile.Value(), 
 										   host, port,
 										   jobQueueDBName?jobQueueDBName:"", 
 										   jobQueueDBUser);
@@ -576,16 +576,16 @@
 	  //const time_t scheddbirthdate = jqDBManager.getScheddbirthdate();
 
   ad->EvalInteger (ATTR_CLUSTER_ID, NULL, cid);
   ad->EvalInteger (ATTR_PROC_ID, NULL, pid);
 
-  scheddbirthdate_str.sprintf("%lu", scheddbirthdate);
+  scheddbirthdate_str.formatstr("%lu", scheddbirthdate);
 
  {
-	  sql_stmt.sprintf(
+	  sql_stmt.formatstr(
 					   "DELETE FROM Jobs_Horizontal_History WHERE scheddname = '%s' AND scheddbirthdate = %lu AND cluster_id = %d AND proc_id = %d", scheddname, (unsigned long)scheddbirthdate, cid, pid);
-	  sql_stmt2.sprintf(
+	  sql_stmt2.formatstr(
 						"INSERT INTO Jobs_Horizontal_History(scheddname, scheddbirthdate, cluster_id, proc_id, enteredhistorytable) VALUES('%s', %lu, %d, %d, current_timestamp)", scheddname, (unsigned long)scheddbirthdate, cid, pid);
 
 	  if (DBObj->execCommand(sql_stmt.Value()) == QUILL_FAILURE) {
 		  dprintf(D_ALWAYS, "Executing Statement --- Error\n");
 		  dprintf(D_ALWAYS, "sql = %s\n", sql_stmt.Value());
@@ -662,11 +662,11 @@
 			  clock = atoi(value.Value());
 
 			  {
 				  ts_expr = condor_ttdb_buildts(&clock, dt);	
 				  
-				  sql_stmt.sprintf(
+				  sql_stmt.formatstr(
 								   "UPDATE Jobs_Horizontal_History SET %s = (%s) WHERE scheddname = '%s' and scheddbirthdate = %lu and cluster_id = %d and proc_id = %d", name.Value(), ts_expr.Value(), scheddname, (unsigned long)scheddbirthdate, cid, pid);
 			  }
 		  }	else {
 				  /* strip double quotes for string values */
 			  if (attr_type == CONDOR_TT_TYPE_CLOB ||
@@ -678,24 +678,24 @@
 			  }
 			  
 			  newvalue = condor_ttdb_fillEscapeCharacters(value.Value(), dt);
 
 			  {
-				  sql_stmt.sprintf( 
+				  sql_stmt.formatstr( 
 								   "UPDATE Jobs_Horizontal_History SET %s = '%s' WHERE scheddname = '%s' and scheddbirthdate = %lu and cluster_id = %d and proc_id = %d", name.Value(), newvalue.Value(), scheddname, (unsigned long)scheddbirthdate, cid, pid);
 			  }
 		  }
 	  } else {
 		  newvalue = condor_ttdb_fillEscapeCharacters(value.Value(), dt);
 		  
 		  {
-			  sql_stmt.sprintf(
+			  sql_stmt.formatstr(
 							   "DELETE FROM Jobs_Vertical_History WHERE scheddname = '%s' AND scheddbirthdate = %lu AND cluster_id = %d AND proc_id = %d AND attr = '%s'", scheddname, (unsigned long)scheddbirthdate, cid, pid, name.Value());
 		  }
 			
 		  {
-			  sql_stmt2.sprintf( 
+			  sql_stmt2.formatstr( 
 								"INSERT INTO Jobs_Vertical_History(scheddname, scheddbirthdate, cluster_id, proc_id, attr, val) VALUES('%s', %lu, %d, %d, '%s', '%s')", scheddname, (unsigned long)scheddbirthdate, cid, pid, name.Value(), newvalue.Value());
 		  }
 	  }
 
 	  {
