diff --git a/src/condor_gridmanager/globusresource.cpp b/src/condor_gridmanager/globusresource.cpp
index e51d583..a72c108 100644
--- a/src/condor_gridmanager/globusresource.cpp
+++ b/src/condor_gridmanager/globusresource.cpp
@@ -780,7 +780,8 @@ GlobusResource::SubmitMonitorJob()
 	contact.sprintf( "%s/jobmanager-fork", resourceName );
 
 	rc = monitorGahp->globus_gram_client_job_request( contact.Value(),
-													  RSL.Value(), 0, 
+													  RSL.Value(),
+													  param_boolean( "DELEGATE_FULL_JOB_GSI_CREDENTIALS", false ) ? 0 : 1, 
 													  monitorGahp->getGt2CallbackContact(),
 													  NULL );
 
