From 2e1f7495a75022b593333ce151e6fdd0f7572a2e Mon Sep 17 00:00:00 2001
From: Quintin Hill <Quintin.Hill@arqiva.com>
Date: Thu, 13 Apr 2017 13:03:43 +0100
Subject: [PATCH 2/3] condor_io: fix build without openssl

---
 src/condor_io/authentication.cpp | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/src/condor_io/authentication.cpp b/src/condor_io/authentication.cpp
index 1036762112..172b5f0830 100644
--- a/src/condor_io/authentication.cpp
+++ b/src/condor_io/authentication.cpp
@@ -1003,7 +1003,12 @@ int Authentication::handshake(MyString my_methods, bool non_blocking) {
 			dprintf (D_SECURITY, "HANDSHAKE: excluding KERBEROS: %s\n", "Initialization failed");
 			method_bitmask &= ~CAUTH_KERBEROS;
 		}
-		if ( (method_bitmask & CAUTH_SSL) && Condor_Auth_SSL::Initialize() == false ) {
+#ifdef HAVE_EXT_OPENSSL
+		if ( (method_bitmask & CAUTH_SSL) && Condor_Auth_SSL::Initialize() == false )
+#else
+		if (method_bitmask & CAUTH_SSL)
+#endif
+		{
 			dprintf (D_SECURITY, "HANDSHAKE: excluding SSL: %s\n", "Initialization failed");
 			method_bitmask &= ~CAUTH_SSL;
 		}
@@ -1058,7 +1063,12 @@ Authentication::handshake_continue(MyString my_methods, bool non_blocking)
 		dprintf (D_SECURITY, "HANDSHAKE: excluding KERBEROS: %s\n", "Initialization failed");
 		shouldUseMethod &= ~CAUTH_KERBEROS;
 	}
-	if ( (shouldUseMethod & CAUTH_SSL) && Condor_Auth_SSL::Initialize() == false ) {
+#ifdef HAVE_EXT_OPENSSL
+	if ( (shouldUseMethod & CAUTH_SSL) && Condor_Auth_SSL::Initialize() == false )
+#else
+	if (shouldUseMethod & CAUTH_SSL)
+#endif
+	{
 		dprintf (D_SECURITY, "HANDSHAKE: excluding SSL: %s\n", "Initialization failed");
 		shouldUseMethod &= ~CAUTH_SSL;
 	}
-- 
2.13.5

