From 6166fbca46d4c236a27413429a1e74095899a8c5 Mon Sep 17 00:00:00 2001
From: Lukas Slebodnik <slebodnik@sors.com>
Date: Mon, 17 Oct 2011 08:54:46 +0200
Subject: [PATCH] Fix cmake ethtool detection.

---
 build/cmake/CondorConfigure.cmake |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/build/cmake/CondorConfigure.cmake b/build/cmake/CondorConfigure.cmake
index fdbb1ed..c0d2518 100644
--- a/build/cmake/CondorConfigure.cmake
+++ b/build/cmake/CondorConfigure.cmake
@@ -310,12 +310,12 @@ elseif(${OS_NAME} STREQUAL "LINUX")
 	set(DOES_SAVE_SIGSTATE ON)
 	check_symbol_exists(SIOCETHTOOL "linux/sockios.h" HAVE_DECL_SIOCETHTOOL)
 	check_symbol_exists(SIOCGIFCONF "linux/sockios.h" HAVE_DECL_SIOCGIFCONF)
-	check_include_files("linux/ethtool.h" HAVE_LINUX_ETHTOOL_H)
+	check_include_files("linux/types.h" HAVE_LINUX_TYPES_H)
+	check_include_files("linux/types.h;linux/ethtool.h" HAVE_LINUX_ETHTOOL_H)
 	check_include_files("linux/magic.h" HAVE_LINUX_MAGIC_H)
 	check_include_files("linux/nfsd/const.h" HAVE_LINUX_NFSD_CONST_H)
 	check_include_files("linux/personality.h" HAVE_LINUX_PERSONALITY_H)
 	check_include_files("linux/sockios.h" HAVE_LINUX_SOCKIOS_H)
-	check_include_files("linux/types.h" HAVE_LINUX_TYPES_H)
 
 	find_library(HAVE_X11 X11)
 	dprint("Threaded functionality only enable in Linux and Windows")
-- 
1.7.4.1

