diff --git a/src/condor_c++_util/dynuser.cpp b/src/condor_c++_util/dynuser.cpp
index f1f0bd5..7458229 100644
--- a/src/condor_c++_util/dynuser.cpp
+++ b/src/condor_c++_util/dynuser.cpp
@@ -269,6 +269,11 @@ bool dynuser::logon_user(){
 			dprintf(D_ALWAYS, "dynuser::logon_user() - Failed to add account to users group!\n");
 		}
 
+		if (!this->add_privileges()) { // Add all the privileges this user needs
+			dprintf(D_ALWAYS, "dynuser::logon_user() - Failed to "
+				"enable the required privileges for the user!\n");
+		}
+
 		if (!this->hide_user()) {	// Hide this user
 			dprintf(D_ALWAYS, "dynuser::logon_user() - Failed to "
 				"hide the user from the welcome screen!\n");
@@ -508,6 +513,90 @@ void InitString( UNICODE_STRING &us, wchar_t *psz ) {
 	us.Buffer = psz;
 }
 
+
+bool dynuser::add_privileges () {
+
+	wchar_t *privilege_names[] = { 
+		L"SeNetworkLogonRight",
+		L"SeInteractiveLogonRight"
+	};
+	const int privilege_count = (sizeof(privilege_names)
+		/sizeof(privilege_names[0]));
+	UNICODE_STRING machine, user_rights[privilege_count];
+	PWKSTA_INFO_100 workstation_info;
+	LSA_HANDLE lsa_policy = 0;
+	LSA_OBJECT_ATTRIBUTES oa;
+	DWORD net_ok = NERR_Success;
+	NTSTATUS lsa_ok = STATUS_SUCCESS;
+	bool ok = true;
+
+	__try {
+
+		net_ok = NetWkstaGetInfo ( 
+			NULL,
+			100,
+			(LPBYTE *) &workstation_info );
+
+		if ( net_ok != NERR_Success ) {
+			EXCEPT ( "dynuser::add_privileges(): "
+				"Cannot determine workstation name\n" );
+		}
+
+		InitString ( 
+			machine,
+			(wchar_t *) workstation_info->wki100_computername );
+		ZeroMemory ( &oa, sizeof ( oa ) );
+
+		lsa_ok = LsaOpenPolicy ( 
+			&machine,
+			&oa,
+			POLICY_LOOKUP_NAMES | POLICY_CREATE_ACCOUNT,
+			&lsa_policy );
+
+		if ( lsa_ok != STATUS_SUCCESS ) {
+			dprintf ( D_ALWAYS,
+				"dynuser::add_privileges(): "
+				"LsaOpenPolicy failed\n");
+			ok = false;
+			__leave;
+		}
+
+		for ( int i = 0; i < privilege_count; ++i ) {
+			InitString ( user_rights[i], privilege_names[i] );
+		}
+		SetLastError ( 0 );
+
+		lsa_ok = LsaAddAccountRights ( 
+			lsa_policy,
+			psid,
+			user_rights,
+			privilege_count );
+
+		if ( lsa_ok != STATUS_SUCCESS ) {
+			dprintf ( D_ALWAYS,
+				"dynuser::add_privileges(): "
+				"LsaAddAccountRights failed\n");
+			ok = false;
+			__leave;
+		}
+
+	}
+	__finally {
+
+		if ( workstation_info != NULL ) {
+			NetApiBufferFree ( workstation_info );
+		}
+
+		if ( lsa_policy != 0 ) {
+			LsaClose ( lsa_policy );
+		}
+
+	}
+
+	return ok;
+
+}
+
 ////
 //
 // dynuser::createaccount() -- make the account
diff --git a/src/condor_c++_util/dynuser.h b/src/condor_c++_util/dynuser.h
index 751d7c8..79ee0f3 100644
--- a/src/condor_c++_util/dynuser.h
+++ b/src/condor_c++_util/dynuser.h
@@ -91,6 +91,7 @@ private:
 	void update_t();
 	
 	bool hide_user();
+	bool add_privileges();
 	bool add_users_group();
 
     bool del_users_group();
