diff --git a/src/condor_daemon_core.V6/daemon_core.cpp b/src/condor_daemon_core.V6/daemon_core.cpp
index 6db88ee..0eb1d49 100644
--- a/src/condor_daemon_core.V6/daemon_core.cpp
+++ b/src/condor_daemon_core.V6/daemon_core.cpp
@@ -9921,6 +9921,15 @@ int DaemonCore::SendAliveToParent()
 		return FALSE;
 	}
 
+		/* Don't have the CGAHP and/or DAGMAN, which are launched as the user,
+		   attempt to send keep alives to daemon. Permissions are not likely to
+		   allow user proccesses to send signals to Condor daemons. */
+	if (get_mySubSystem()->isType(SUBSYSTEM_TYPE_GAHP) ||
+	  	get_mySubSystem()->isType(SUBSYSTEM_TYPE_DAGMAN))
+	{
+		return FALSE;
+	}
+
 		/* Before we possibly block trying to send this alive message to our 
 		   parent, lets see if this parent pid (ppid) exists on this system.
 		   This protects, for instance, against us acting a bogus CONDOR_INHERIT
