diff --git a/src/condor_negotiator.V6/Accountant.cpp b/src/condor_negotiator.V6/Accountant.cpp
index 8845ee7..50b851e 100644
--- a/src/condor_negotiator.V6/Accountant.cpp
+++ b/src/condor_negotiator.V6/Accountant.cpp
@@ -309,7 +309,7 @@ void parse_group_name(const string& gname, vector<string>& gpath) {
     gpath.clear();
     string::size_type cur = 0;
     while (true) {
-        string::size_type nxt = gname.find_first_of('.', cur);
+        string::size_type nxt = gname.find_first_of(':', cur);
         string::size_type n = (nxt == string::npos) ? string::npos : nxt-cur;
         gpath.push_back(gname.substr(cur, n));
         if (nxt == string::npos) break;
@@ -346,7 +346,7 @@ GroupEntry* Accountant::GetAssignedGroup(const MyString& CustomerName, bool& IsG
     string gname=subname.substr(0, pos);
 
     // is there a group/user separator?
-    pos = gname.find_last_of('.');
+    pos = gname.find_last_of(':');
     if (pos != string::npos) {
         // everything prior to separator is group name
         gname = gname.substr(0, pos);
