diff --git a/src/condor_utils/my_popen.cpp b/src/condor_utils/my_popen.cpp
index 4ea6b56..870caac 100644
--- a/src/condor_utils/my_popen.cpp
+++ b/src/condor_utils/my_popen.cpp
@@ -469,6 +469,12 @@ my_popenv_impl( const char *const args[],
 		close(pipe_d2[0]);
 		close(pipe_d[0]);
 		close(pipe_d[1]);
+
+		/* Wait for child process to exit */
+		while( waitpid(pid,NULL,0) < 0 && errno == EINTR ) {
+			/* NOOP */
+		}
+
 		return NULL;
 	}
 		/* Handle case where exec fails */
@@ -476,6 +482,12 @@ my_popenv_impl( const char *const args[],
 		fclose(fh);
 		close(pipe_d[0]);
 		close(pipe_d[1]);
+
+		/* Wait for child process to exit */
+		while( waitpid(pid,NULL,0) < 0 && errno == EINTR ) {
+			/* NOOP */
+		}
+
 		errno = exit_code;
 		return NULL;
 	}
