diff --git a/src/job_registry.c b/src/job_registry.c
index 4551597..e65c6b9 100644
--- a/src/job_registry.c
+++ b/src/job_registry.c
@@ -294,7 +294,7 @@ job_registry_purge(const char *path, time_t oldest_creation_date,
      }
     if (ret == 0) break;
 
-    if (cur.cdate < oldest_creation_date)
+    if (cur.cdate < oldest_creation_date && (cur.status != RUNNING && cur.status != IDLE && cur.status != HELD))
      {
       job_registry_unlink_proxy(jra, &cur);
       continue;

