diff --git a/src/condor_gridmanager/proxymanager.cpp b/src/condor_gridmanager/proxymanager.cpp
index 4b6b00e..dbe94ae 100644
--- a/src/condor_gridmanager/proxymanager.cpp
+++ b/src/condor_gridmanager/proxymanager.cpp
@@ -39,6 +39,8 @@
 #include "gridmanager.h"
 #include "condor_string.h"
 
+#include <sstream>
+
 #define HASH_TABLE_SIZE			500
 
 
@@ -273,6 +275,19 @@ AcquireProxy( const ClassAd *job_ad, std::string &error,
 		return NULL;
 	}
 
+	// If Condor-C submitted the job, the proxy_path is relative to the
+	// spool directory.  For the purposes of this function, extend the
+	// proxy path with the ATTR_JOB_IWD
+	if (proxy_path[0] != DIR_DELIM_CHAR) {
+		std::string iwd;
+		job_ad->LookupString(ATTR_JOB_IWD, iwd);
+		if (!iwd.empty()) {
+			std::stringstream ss;
+			ss << iwd << DIR_DELIM_CHAR << proxy_path;
+			proxy_path = ss.str();
+		}   
+	}   
+
 	if ( ProxiesByFilename.lookup( HashKey(proxy_path.c_str()), proxy ) == 0 ) {
 		// We already know about this proxy,
 		// use the existing Proxy struct
