diff --git a/src/condor_tools/CMakeLists.txt b/src/condor_tools/CMakeLists.txt
index 30f3e82..bd41ab9 100644
--- a/src/condor_tools/CMakeLists.txt
+++ b/src/condor_tools/CMakeLists.txt
@@ -75,6 +75,7 @@ if (NOT WINDOWS)
 			message( FATAL_ERROR "Can't find libdl.a, needed for condor_ckpt_probe" )
 		endif()
 		add_executable( condor_ckpt_probe probe_process.cpp ../condor_ckpt/maps.c ../condor_utils/condor_pers.cpp ../safefile/safe_open.cpp )
+		install( TARGETS condor_ckpt_probe DESTINATION "${C_LIBEXEC}" )
         set_property( TARGET condor_ckpt_probe PROPERTY COMPILE_FLAGS -static )
         set_property( TARGET condor_ckpt_probe PROPERTY LINK_FLAGS -static )
 	endif(STD_UNIVERSE)
