diff --git a/src/classad_analysis/analysis.h b/src/classad_analysis/analysis.h
index c6f38cf..27b5943 100644
--- a/src/classad_analysis/analysis.h
+++ b/src/classad_analysis/analysis.h
@@ -24,7 +24,7 @@
 //#define EXPERIMENTAL
 
 #define WANT_CLASSAD_NAMESPACE
-#include "condor_fix_iostream.h"
+#include <iostream>
 #include "classad/classad_distribution.h"
 #include "condor_classad.h"
 #include "condor_attributes.h"
diff --git a/src/classad_analysis/conversion.h b/src/classad_analysis/conversion.h
index 11daf6e..8630f1f 100644
--- a/src/classad_analysis/conversion.h
+++ b/src/classad_analysis/conversion.h
@@ -22,7 +22,7 @@
 #define __CONVERSION_H__
 
 #define WANT_CLASSAD_NAMESPACE
-#include "condor_fix_iostream.h"
+#include <iostream>
 #include "classad/classad_distribution.h"
 #include "condor_classad.h"
 
diff --git a/src/classad_analysis/explain.h b/src/classad_analysis/explain.h
index 5d3051c..926ec93 100644
--- a/src/classad_analysis/explain.h
+++ b/src/classad_analysis/explain.h
@@ -22,7 +22,7 @@
 #define __EXPLAIN_H__
 
 #define WANT_CLASSAD_NAMESPACE
-#include "condor_fix_iostream.h"
+#include <iostream>
 #include "classad/classad_distribution.h"
 #include "list.h"
 #include "interval.h"
diff --git a/src/classad_analysis/interval.h b/src/classad_analysis/interval.h
index 9831300..b5a3ce6 100644
--- a/src/classad_analysis/interval.h
+++ b/src/classad_analysis/interval.h
@@ -22,7 +22,7 @@
 #define __INTERVAL_H__
 
 #define WANT_CLASSAD_NAMESPACE
-#include "condor_fix_iostream.h"
+#include <iostream>
 #include "classad/classad_distribution.h"
 #include "list.h"
 #include "extArray.h"
diff --git a/src/classad_analysis/resourceGroup.h b/src/classad_analysis/resourceGroup.h
index 3c601e4..b96e65a 100644
--- a/src/classad_analysis/resourceGroup.h
+++ b/src/classad_analysis/resourceGroup.h
@@ -22,7 +22,7 @@
 #define __RESOURCEGROUP_H__
 
 #define WANT_CLASSAD_NAMESPACE
-#include "condor_fix_iostream.h"
+#include <iostream>
 #include "classad/classad_distribution.h"
 #include "list.h"
 
diff --git a/src/condor_c++_util/MapFile.cpp b/src/condor_c++_util/MapFile.cpp
index fafdc40..42a5257 100644
--- a/src/condor_c++_util/MapFile.cpp
+++ b/src/condor_c++_util/MapFile.cpp
@@ -18,6 +18,7 @@
  ***************************************************************/
 
 #include "condor_common.h"
+#include "condor_open.h"
 #include "condor_debug.h"
 #include "MapFile.h"
 
diff --git a/src/condor_c++_util/Queue.t.cpp b/src/condor_c++_util/Queue.t.cpp
index 5776f07..5366941 100644
--- a/src/condor_c++_util/Queue.t.cpp
+++ b/src/condor_c++_util/Queue.t.cpp
@@ -17,7 +17,7 @@
  *
  ***************************************************************/
 
-#include <iostream.h>
+#include <iostream>
 #include <stdio.h>
 #include "Queue.h"
 
diff --git a/src/condor_c++_util/credential.h b/src/condor_c++_util/credential.h
index 5c8ea49..ed8d4b4 100644
--- a/src/condor_c++_util/credential.h
+++ b/src/condor_c++_util/credential.h
@@ -32,7 +32,7 @@
 #ifndef WANT_CLASSAD_NAMESPACE
 #define WANT_CLASSAD_NAMESPACE
 #endif
-#include "condor_fix_iostream.h"
+#include <iostream>
 #include "classad/classad_distribution.h"
 
 #include "MyString.h"
diff --git a/src/condor_c++_util/directory.cpp b/src/condor_c++_util/directory.cpp
index 8bcf21f..7a04fa7 100644
--- a/src/condor_c++_util/directory.cpp
+++ b/src/condor_c++_util/directory.cpp
@@ -19,6 +19,7 @@
 
 
 #include "condor_common.h"
+#include "condor_open.h"
 #include "condor_constants.h"
 #include "condor_debug.h"
 #include "directory.h"
diff --git a/src/condor_c++_util/file_url.cpp b/src/condor_c++_util/file_url.cpp
index c07934e..a1bd7db 100644
--- a/src/condor_c++_util/file_url.cpp
+++ b/src/condor_c++_util/file_url.cpp
@@ -20,6 +20,7 @@
 
  
 #include "condor_common.h"
+#include "condor_open.h"
 #include "url_condor.h"
 
 static
diff --git a/src/condor_c++_util/memory_file.cpp b/src/condor_c++_util/memory_file.cpp
index c618238..f58ad0c 100644
--- a/src/condor_c++_util/memory_file.cpp
+++ b/src/condor_c++_util/memory_file.cpp
@@ -26,7 +26,8 @@
 #endif
 
 #include "memory_file.h"
-#include "condor_fix_iostream.h"
+#include <iostream>
+using namespace std;
 
 static const int DEFAULT_BUFFER_SIZE=1024;
 static const int COMPARE_BUFFER_SIZE=10000;   
diff --git a/src/condor_c++_util/read_multiple_logs.cpp b/src/condor_c++_util/read_multiple_logs.cpp
index 7d793e7..f6bd080 100644
--- a/src/condor_c++_util/read_multiple_logs.cpp
+++ b/src/condor_c++_util/read_multiple_logs.cpp
@@ -28,7 +28,7 @@
 #ifndef WANT_CLASSAD_NAMESPACE
 #define WANT_CLASSAD_NAMESPACE
 #endif
-#include "condor_fix_iostream.h"
+#include <iostream>
 #include "classad/classad_distribution.h"
 #endif
 #include "fs_util.h"
diff --git a/src/condor_c++_util/read_user_log.cpp b/src/condor_c++_util/read_user_log.cpp
index 43dee98..8830eac 100644
--- a/src/condor_c++_util/read_user_log.cpp
+++ b/src/condor_c++_util/read_user_log.cpp
@@ -18,6 +18,7 @@
  ***************************************************************/
 
 #include "condor_common.h"
+#include "condor_open.h"
 #include "condor_debug.h"
 #include <stdarg.h>
 #include "read_user_log.h"
diff --git a/src/condor_c++_util/test_mapfile.cpp b/src/condor_c++_util/test_mapfile.cpp
index 52b90f1..5b1aba9 100644
--- a/src/condor_c++_util/test_mapfile.cpp
+++ b/src/condor_c++_util/test_mapfile.cpp
@@ -21,7 +21,7 @@
 #include "MyString.h"
 #include "MapFile.h"
 
-#include "condor_fix_iostream.h"
+#include <iostream>
 
 
 /*
diff --git a/src/condor_c++_util/test_queue.cpp b/src/condor_c++_util/test_queue.cpp
index 0012bad..4439e3f 100644
--- a/src/condor_c++_util/test_queue.cpp
+++ b/src/condor_c++_util/test_queue.cpp
@@ -17,7 +17,7 @@
  *
  ***************************************************************/
 
-#include <iostream.h>
+#include <iostream>
 #include "queue.h"
 
 int main (int argc, char *argv[]) {
diff --git a/src/condor_c++_util/test_simplelist.cpp b/src/condor_c++_util/test_simplelist.cpp
index 2733ed1..ad69aa4 100644
--- a/src/condor_c++_util/test_simplelist.cpp
+++ b/src/condor_c++_util/test_simplelist.cpp
@@ -17,7 +17,7 @@
  *
  ***************************************************************/
 
-#include <iostream.h>
+#include <iostream>
 #include "simplelist.h"
 
 void iterate (const SimpleList<int> & list) {
diff --git a/src/condor_c++_util/test_stringSpace.cpp b/src/condor_c++_util/test_stringSpace.cpp
index 173c12a..2abb995 100644
--- a/src/condor_c++_util/test_stringSpace.cpp
+++ b/src/condor_c++_util/test_stringSpace.cpp
@@ -17,7 +17,7 @@
  *
  ***************************************************************/
 
-#include <iostream.h>
+#include <iostream>
 #include "stringSpace.h"
 
 /*-------------------------------------------------------------
diff --git a/src/condor_ckpt_server/fileindex2.cpp b/src/condor_ckpt_server/fileindex2.cpp
index 4c726bf..c3975b4 100644
--- a/src/condor_ckpt_server/fileindex2.cpp
+++ b/src/condor_ckpt_server/fileindex2.cpp
@@ -23,7 +23,8 @@
 #include "fileindex2.h"
 #include "gen_lib.h" 
 
-#include "condor_fix_iostream.h"
+#include <iostream>
+using namespace std;
 
 FileIndex::FileIndex()
 {
diff --git a/src/condor_ckpt_server/fileinfo2.cpp b/src/condor_ckpt_server/fileinfo2.cpp
index 91fdd1e..6abd452 100644
--- a/src/condor_ckpt_server/fileinfo2.cpp
+++ b/src/condor_ckpt_server/fileinfo2.cpp
@@ -23,9 +23,10 @@
 #include "fileinfo2.h"
 #include "network2.h"
 #include "gen_lib.h"
-#include "condor_fix_iostream.h"
-#include "condor_fix_fstream.h"
-#include "condor_fix_iomanip.h"
+#include <iostream>
+#include <fstream>
+#include <iomanip>
+using namespace std;
 
 FileInformation::FileInformation()
 {
diff --git a/src/condor_ckpt_server/protocol.cpp b/src/condor_ckpt_server/protocol.cpp
index 5c51a48..ad1fec5 100644
--- a/src/condor_ckpt_server/protocol.cpp
+++ b/src/condor_ckpt_server/protocol.cpp
@@ -1,7 +1,7 @@
 #include "condor_common.h"
 #include "condor_debug.h"
-#include "condor_fix_iostream.h"
-#include "condor_fix_fstream.h"
+#include <iostream>
+#include <fstream>
 #include "MyString.h"
 
 #include "constants2.h"
diff --git a/src/condor_ckpt_server/server2.cpp b/src/condor_ckpt_server/server2.cpp
index 3fb5ab6..35d5c30 100644
--- a/src/condor_ckpt_server/server2.cpp
+++ b/src/condor_ckpt_server/server2.cpp
@@ -30,8 +30,9 @@
 #include "condor_socket_types.h"
 #include "subsystem_info.h"
 #include "condor_netdb.h"
-#include "condor_fix_iostream.h"
-#include "condor_fix_fstream.h"
+#include <iostream>
+#include <fstream>
+#include "condor_open.h"
 #include "directory.h"
 #include "basename.h"
 #include "stat_info.h"
@@ -45,6 +46,7 @@
 #include "classad_collection.h"
 #include "daemon.h"
 #include "protocol.h"
+using namespace std;
 
 XferSummary	xfer_summary;
 
diff --git a/src/condor_ckpt_server/server2.h b/src/condor_ckpt_server/server2.h
index b3e9d53..65aa330 100644
--- a/src/condor_ckpt_server/server2.h
+++ b/src/condor_ckpt_server/server2.h
@@ -53,7 +53,7 @@ class Server
     int            max_req_sd_plus1;
     IMDS           imds;
     TransferState  transfers;
-    ofstream       log_file;
+    std::ofstream       log_file;
     int            num_store_xfers;
     int            num_restore_xfers;
 	int            num_replicate_xfers;
diff --git a/src/condor_ckpt_server/server_filestat.cpp b/src/condor_ckpt_server/server_filestat.cpp
index 4481a64..02ee4d1 100644
--- a/src/condor_ckpt_server/server_filestat.cpp
+++ b/src/condor_ckpt_server/server_filestat.cpp
@@ -47,7 +47,7 @@
 #include "server_network.h"
 #include <stdlib.h>
 #include <string.h>
-#include <iostream.h>
+#include <iostream>
 #include <iomanip.h>
 #include "server_filestat.h"
 #include <stdio.h>
diff --git a/src/condor_ckpt_server/signal2.cpp b/src/condor_ckpt_server/signal2.cpp
index 375963d..7348c7a 100644
--- a/src/condor_ckpt_server/signal2.cpp
+++ b/src/condor_ckpt_server/signal2.cpp
@@ -22,7 +22,7 @@
 #include "condor_debug.h"
 #include "signal2.h"
 #include "constants2.h"
-#include "condor_fix_iostream.h"
+#include <iostream>
 
 
 void BlockSignal(int sig)
diff --git a/src/condor_ckpt_server/xferstat2.cpp b/src/condor_ckpt_server/xferstat2.cpp
index a6bdeab..cc6867a 100644
--- a/src/condor_ckpt_server/xferstat2.cpp
+++ b/src/condor_ckpt_server/xferstat2.cpp
@@ -49,7 +49,8 @@
 #include "xferstat2.h"
 #include "xfer_summary.h"
 
-#include "condor_fix_iostream.h"
+#include <iostream>
+using namespace std;
 
 /* Class TransferState ********************************************************
 *                                                                             *
diff --git a/src/condor_credd/classadUtil.h b/src/condor_credd/classadUtil.h
index 7861f8a..043702f 100644
--- a/src/condor_credd/classadUtil.h
+++ b/src/condor_credd/classadUtil.h
@@ -26,7 +26,7 @@
 #ifndef WANT_CLASSAD_NAMESPACE
 #define WANT_CLASSAD_NAMESPACE
 #endif
-#include "condor_fix_iostream.h"
+#include <iostream>
 #include "classad/classad_distribution.h"
 
 using namespace std;
diff --git a/src/condor_daemon_client/daemon.cpp b/src/condor_daemon_client/daemon.cpp
index c4d6c11..6c1fd51 100644
--- a/src/condor_daemon_client/daemon.cpp
+++ b/src/condor_daemon_client/daemon.cpp
@@ -22,6 +22,7 @@
 #include "condor_debug.h"
 #include "condor_config.h"
 #include "condor_ver_info.h"
+#include "condor_open.h"
 
 #include "daemon.h"
 #include "condor_string.h"
diff --git a/src/condor_daemon_client/dc_starter.cpp b/src/condor_daemon_client/dc_starter.cpp
index d62c84f..8ae4f66 100644
--- a/src/condor_daemon_client/dc_starter.cpp
+++ b/src/condor_daemon_client/dc_starter.cpp
@@ -24,6 +24,7 @@
 #include "condor_string.h"
 #include "condor_ver_info.h"
 #include "condor_attributes.h"
+#include "condor_open.h"
 
 #include "daemon.h"
 #include "dc_starter.h"
diff --git a/src/condor_daemon_core.V6/daemon_core.cpp b/src/condor_daemon_core.V6/daemon_core.cpp
index 88127f0..4dd3b8d 100644
--- a/src/condor_daemon_core.V6/daemon_core.cpp
+++ b/src/condor_daemon_core.V6/daemon_core.cpp
@@ -107,6 +107,7 @@ CRITICAL_SECTION Big_fat_mutex; // coarse grained mutex for debugging purposes
 #include "basename.h"
 #include "condor_threads.h"
 #include "shared_port_endpoint.h"
+#include "condor_open.h"
 
 #include "valgrind.h"
 
diff --git a/src/condor_daemon_core.V6/daemon_core_main.cpp b/src/condor_daemon_core.V6/daemon_core_main.cpp
index 6ee040f..2651fb3 100644
--- a/src/condor_daemon_core.V6/daemon_core_main.cpp
+++ b/src/condor_daemon_core.V6/daemon_core_main.cpp
@@ -19,6 +19,7 @@
 
 
 #include "condor_common.h"
+#include "condor_open.h"
 #include "condor_config.h"
 #include "util_lib_proto.h"
 #include "basename.h"
@@ -703,7 +704,7 @@ install_core_dump_handler()
 
 void
 drop_core_in_log( void )
-{
+{
 	// chdir to the LOG directory so that if we dump a core
 	// it will go there.
 	// and on Win32, tell our ExceptionHandler class to drop
@@ -740,9 +741,9 @@ drop_core_in_log( void )
 		if ( binpath ) {
 			SetEnv( "_NT_SYMBOL_PATH", binpath );
 			free(binpath);
-		}
-
-		// give the handler our pid
+		}
+
+		// give the handler our pid
 		g_ExceptionHandler.SetPID ( daemonCore->getpid () );
 	}
 #endif
diff --git a/src/condor_daemon_core.V6/soap_core.cpp b/src/condor_daemon_core.V6/soap_core.cpp
index 3830356..bba55b8 100644
--- a/src/condor_daemon_core.V6/soap_core.cpp
+++ b/src/condor_daemon_core.V6/soap_core.cpp
@@ -30,6 +30,7 @@
 #include "directory.h"
 #include "stdsoap2.h"
 #include "soap_core.h"
+#include "condor_open.h"
 
 #include "mimetypes.h"
 
diff --git a/src/condor_daemon_core.V6/stdsoap2.h.patch b/src/condor_daemon_core.V6/stdsoap2.h.patch
index a0fe3ed..e463427 100644
--- a/src/condor_daemon_core.V6/stdsoap2.h.patch
+++ b/src/condor_daemon_core.V6/stdsoap2.h.patch
@@ -44,17 +44,6 @@
  #  define __USE_STD_IOSTREAM
  #  define HAVE_WCTOMB
  #  define HAVE_MBTOWC
-@@ -418,8 +427,8 @@
- #endif
- 
- #if defined(__cplusplus) && !defined(WITH_LEAN)
--# include <string>
--# include <iostream>
-+# include <string.h>
-+# include "condor_fix_iostream.h"
- #endif
- 
- #ifdef WITH_NOHTTP
 @@ -571,7 +562,7 @@
  # define SOAP_SOCKLEN_T SOCKLEN_T
  #elif defined(__socklen_t_defined) || defined(_SOCKLEN_T) || defined(CYGWIN) || defined(FREEBSD) || defined(__FreeBSD__) || defined(__QNX__) || defined(QNX) || defined(_AIX)
diff --git a/src/condor_had/HAD.cpp b/src/condor_had/HAD.cpp
index 2e71f11..c569eb4 100644
--- a/src/condor_had/HAD.cpp
+++ b/src/condor_had/HAD.cpp
@@ -39,7 +39,8 @@
 	#include "ReplicaStateMachine.h"
 #endif // USE_REPLICATION
 
-#include "condor_fix_iostream.h"
+#include <iostream>
+using namespace std;
 
 extern "C" int SetSyscalls(int val){return val;}
 extern char* myName;
diff --git a/src/condor_had/HAD_Version.cpp b/src/condor_had/HAD_Version.cpp
index 8abea67..e6d324d 100644
--- a/src/condor_had/HAD_Version.cpp
+++ b/src/condor_had/HAD_Version.cpp
@@ -29,7 +29,8 @@
 
 #include "Version.h"
 #include "FilesOperations.h"
-#include "condor_fix_fstream.h"
+#include <fstream>
+using namespace std;
 
 time_t Version::m_lastModifiedTime = -1;
 
diff --git a/src/condor_had/Utils.cpp b/src/condor_had/Utils.cpp
index 5b58ca0..459b2b5 100644
--- a/src/condor_had/Utils.cpp
+++ b/src/condor_had/Utils.cpp
@@ -33,7 +33,9 @@
 #endif
 #include "condor_netdb.h"
 
-#include "condor_fix_fstream.h"
+#include <fstream>
+
+using namespace std;
 
 // for MD5 blocks computation
 #define FILE_CHUNK_SIZE               (100)
diff --git a/src/condor_includes/condor_common.h b/src/condor_includes/condor_common.h
index c081bc8..7cd2b04 100644
--- a/src/condor_includes/condor_common.h
+++ b/src/condor_includes/condor_common.h
@@ -43,6 +43,7 @@
 **********************************************************************/
 #include "condor_constants.h"
 #include "condor_macros.h"
+#include "condor_open.h"
 #include "condor_blkng_full_disk_io.h"
 #include "condor_create_id.h"
 
diff --git a/src/condor_includes/condor_fix_fstream.h b/src/condor_includes/condor_fix_fstream.h
deleted file mode 100644
index d0243d4..0000000
--- a/src/condor_includes/condor_fix_fstream.h
+++ /dev/null
@@ -1,48 +0,0 @@
-
-/***************************************************************
- *
- * Copyright (C) 1990-2007, Condor Team, Computer Sciences Department,
- * University of Wisconsin-Madison, WI.
- * 
- * Licensed under the Apache License, Version 2.0 (the "License"); you
- * may not use this file except in compliance with the License.  You may
- * obtain a copy of the License at
- * 
- *    http://www.apache.org/licenses/LICENSE-2.0
- * 
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- *
- ***************************************************************/
-
-#if !defined(__CONDOR_FIX_FSTREAM_H) && defined(__cplusplus)
-#define __CONDOR_FIX_FSTREAM_H
-
-	/* Save what we redefined open() and fopen() to be, if anything.
- 	*/
-#ifndef _CONDOR_ALLOW_OPEN
-#	define _CONDOR_OPEN_SAVE open
-#	undef open
-#endif
-#ifndef _CONDOR_ALLOW_FOPEN
-#	define _CONDOR_FOPEN_SAVE fopen
-#	undef fopen
-#endif
-
-#include <fstream>
-using namespace std;
-
-	/* Restore the open() and fopen() macros back to what they were
-	 */
-#ifdef _CONDOR_OPEN_SAVE
-#	define open _CONDOR_OPEN_SAVE
-#endif
-#ifdef _CONDOR_FOPEN_SAVE
-#	define fopen _CONDOR_FOPEN_SAVE
-#endif
-
-
-#endif
diff --git a/src/condor_includes/condor_fix_iomanip.h b/src/condor_includes/condor_fix_iomanip.h
deleted file mode 100644
index 9c6be79..0000000
--- a/src/condor_includes/condor_fix_iomanip.h
+++ /dev/null
@@ -1,48 +0,0 @@
-
-/***************************************************************
- *
- * Copyright (C) 1990-2007, Condor Team, Computer Sciences Department,
- * University of Wisconsin-Madison, WI.
- * 
- * Licensed under the Apache License, Version 2.0 (the "License"); you
- * may not use this file except in compliance with the License.  You may
- * obtain a copy of the License at
- * 
- *    http://www.apache.org/licenses/LICENSE-2.0
- * 
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- *
- ***************************************************************/
-
-#if !defined(__CONDOR_FIX_CLASSAD_DISTRIBUTION_H) && defined(__cplusplus)
-#define __CONDOR_FIX_CLASSAD_DISTRIBUTION_H
-
-	/* Save what we redefined open() and fopen() to be, if anything.
- 	*/
-#ifndef _CONDOR_ALLOW_OPEN
-#	define _CONDOR_OPEN_SAVE open
-#	undef open
-#endif
-#ifndef _CONDOR_ALLOW_FOPEN
-#	define _CONDOR_FOPEN_SAVE fopen
-#	undef fopen
-#endif
-
-#include <iomanip>
-using namespace std;
-
-	/* Restore the open() and fopen() macros back to what they were
-	 */
-#ifdef _CONDOR_OPEN_SAVE
-#	define open _CONDOR_OPEN_SAVE
-#endif
-#ifdef _CONDOR_FOPEN_SAVE
-#	define fopen _CONDOR_FOPEN_SAVE
-#endif
-
-
-#endif
diff --git a/src/condor_includes/condor_fix_iostream.h b/src/condor_includes/condor_fix_iostream.h
deleted file mode 100644
index 9ab9fdb..0000000
--- a/src/condor_includes/condor_fix_iostream.h
+++ /dev/null
@@ -1,48 +0,0 @@
-
-/***************************************************************
- *
- * Copyright (C) 1990-2007, Condor Team, Computer Sciences Department,
- * University of Wisconsin-Madison, WI.
- * 
- * Licensed under the Apache License, Version 2.0 (the "License"); you
- * may not use this file except in compliance with the License.  You may
- * obtain a copy of the License at
- * 
- *    http://www.apache.org/licenses/LICENSE-2.0
- * 
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- *
- ***************************************************************/
-
-#if !defined(__CONDOR_FIX_IOSTREAM_H) && defined(__cplusplus)
-#define __CONDOR_FIX_IOSTREAM_H
-
-	/* Save what we redefined open() and fopen() to be, if anything.
- 	*/
-#ifndef _CONDOR_ALLOW_OPEN
-#	define _CONDOR_OPEN_SAVE open
-#	undef open
-#endif
-#ifndef _CONDOR_ALLOW_FOPEN
-#	define _CONDOR_FOPEN_SAVE fopen
-#	undef fopen
-#endif
-
-#include <iostream>
-using namespace std;
-
-	/* Restore the open() and fopen() macros back to what they were
-	 */
-#ifdef _CONDOR_OPEN_SAVE
-#	define open _CONDOR_OPEN_SAVE
-#endif
-#ifdef _CONDOR_FOPEN_SAVE
-#	define fopen _CONDOR_FOPEN_SAVE
-#endif
-
-
-#endif
diff --git a/src/condor_includes/condor_macros.h b/src/condor_includes/condor_macros.h
index 32df959..d0681ec 100644
--- a/src/condor_includes/condor_macros.h
+++ b/src/condor_includes/condor_macros.h
@@ -42,17 +42,17 @@
 	/* Don't prevent calls to open() or fopen() in the ckpt or 
 	 * remote syscalls code.
 	 */
-#if defined(IN_CKPT_LIB) || defined(REMOTE_SYSCALLS)
+/*#if defined(IN_CKPT_LIB) || defined(REMOTE_SYSCALLS)
 #	if !defined(_CONDOR_ALLOW_OPEN_AND_FOPEN)
 #		define _CONDOR_ALLOW_OPEN_AND_FOPEN 1
 #	endif
-#endif
+#endif*/
 
 	/* For security purposes, do not allow open() or fopen().
 	 * Folks should use the wrapper functions in 
 	 * src/condor_c++_util/condor_open.[hC] 
 	 */
-#ifdef _CONDOR_ALLOW_OPEN_AND_FOPEN
+/*#ifdef _CONDOR_ALLOW_OPEN_AND_FOPEN
 #  ifndef _CONDOR_ALLOW_OPEN
 #     define _CONDOR_ALLOW_OPEN 1
 #  endif
@@ -77,14 +77,15 @@
 #  ifdef fopen
 #    undef fopen
 #  endif
+*/
 /* Condor's fopen macro does not play well with the new 
    version of the Platform SDKs (see cstdio for details) */
-#if !defined (WIN32)
+/*#if !defined (WIN32)
 #   define fopen (Calls_to_fopen_must_use___safe_fopen_wrapper___instead)   
 #endif
 #  ifdef __GNUC__
 #    pragma GCC poison Calls_to_fopen_must_use___safe_fopen_wrapper___instead
 #  endif
-#endif 
+#endif */
 
 #endif /* CONDOR_MACROS_H */
diff --git a/src/condor_includes/condor_system.h b/src/condor_includes/condor_system.h
index 03549b4..462648c 100644
--- a/src/condor_includes/condor_system.h
+++ b/src/condor_includes/condor_system.h
@@ -84,7 +84,7 @@
 #include "condor_fix_assert.h"
 
 #if defined(WIN32)
-#include "condor_fix_iomanip.h"
+#include <iomanip.h>
 #endif
 
 
diff --git a/src/condor_io/cedar.t.cpp b/src/condor_io/cedar.t.cpp
index 1bab92e..e2ccffb 100644
--- a/src/condor_io/cedar.t.cpp
+++ b/src/condor_io/cedar.t.cpp
@@ -26,12 +26,14 @@
 #include "test.h"
 
 #include <stdio.h>
-#include "condor_fix_iostream.h"
+#include <iostream>
 #include <string.h>
 #include <sys/types.h>
 #include <unistd.h>
 #include <stdio.h>
 
+using namespace std;
+
 int main()
 {
 	SafeSock mySock;
diff --git a/src/condor_master.V6/service.WINDOWS.cpp b/src/condor_master.V6/service.WINDOWS.cpp
index f7c3e15..2beb348 100644
--- a/src/condor_master.V6/service.WINDOWS.cpp
+++ b/src/condor_master.V6/service.WINDOWS.cpp
@@ -36,7 +36,7 @@
 #include "..\condor_daemon_core.V6\condor_daemon_core.h"
 #include "master.h"
 
-#include "condor_fix_iostream.h"
+#include <iostream>
 #include <windows.h>
 #include <stdio.h>
 #include <stdlib.h>
diff --git a/src/condor_negotiator.V6/Accountant.cpp b/src/condor_negotiator.V6/Accountant.cpp
index e746ddb..9872da9 100644
--- a/src/condor_negotiator.V6/Accountant.cpp
+++ b/src/condor_negotiator.V6/Accountant.cpp
@@ -21,7 +21,7 @@
 #include "condor_common.h"
 
 #include <math.h>
-#include "condor_fix_iomanip.h"
+#include <iomanip>
 
 #include "condor_accountant.h"
 #include "condor_debug.h"
diff --git a/src/condor_procd/local_client_test.cpp b/src/condor_procd/local_client_test.cpp
index 6618344..8c720d3 100644
--- a/src/condor_procd/local_client_test.cpp
+++ b/src/condor_procd/local_client_test.cpp
@@ -19,7 +19,7 @@
 
 
 #include "condor_common.h"
-#include "condor_fix_iostream.h"
+#include <iostream>
 #include "condor_debug.h"
 #include "local_client.h"
 
diff --git a/src/condor_q.V6/queue.cpp b/src/condor_q.V6/queue.cpp
index 7d98081..a118b13 100644
--- a/src/condor_q.V6/queue.cpp
+++ b/src/condor_q.V6/queue.cpp
@@ -51,7 +51,7 @@
 #include "condor_version.h"
 #include "subsystem_info.h"
 #include "condor_xml_classads.h"
-
+#include "condor_open.h"
 
 #ifdef WANT_CLASSAD_ANALYSIS
 #include "../classad_analysis/analysis.h"
diff --git a/src/condor_release/examples/fstream.cpp b/src/condor_release/examples/fstream.cpp
index 3723564..b695029 100644
--- a/src/condor_release/examples/fstream.cpp
+++ b/src/condor_release/examples/fstream.cpp
@@ -17,7 +17,7 @@
  *
  ***************************************************************/
 
-#include <iostream.h>
+#include <iostream>
 #include <fstream.h>
 
 int main() {
diff --git a/src/condor_schedd.V6/schedd.cpp b/src/condor_schedd.V6/schedd.cpp
index aaa9fb3..fc05377 100644
--- a/src/condor_schedd.V6/schedd.cpp
+++ b/src/condor_schedd.V6/schedd.cpp
@@ -88,6 +88,7 @@
 #include "setenv.h"
 #include "classadHistory.h"
 #include "forkwork.h"
+#include "condor_open.h"
 
 #if HAVE_DLOPEN
 #include "ScheddPlugin.h"
diff --git a/src/condor_schedd.V6/schedd_api.cpp b/src/condor_schedd.V6/schedd_api.cpp
index f1ade1c..c7613df 100644
--- a/src/condor_schedd.V6/schedd_api.cpp
+++ b/src/condor_schedd.V6/schedd_api.cpp
@@ -32,6 +32,7 @@
 
 #include "condor_ckpt_name.h"
 #include "condor_config.h"
+#include "condor_open.h"
 
 #include "schedd_api.h"
 
diff --git a/src/stork/dap_classad_reader.h b/src/stork/dap_classad_reader.h
index 906b22c..86f5952 100644
--- a/src/stork/dap_classad_reader.h
+++ b/src/stork/dap_classad_reader.h
@@ -27,7 +27,7 @@
 #include "dap_constants.h"
 
 #define WANT_CLASSAD_NAMESPACE
-#include "condor_fix_iostream.h"
+#include <iostream>
 #include "classad/classad_distribution.h"
 
 class ClassAd_Reader{
diff --git a/src/stork/dap_client_interface.h b/src/stork/dap_client_interface.h
index c70e590..ef5c7da 100644
--- a/src/stork/dap_client_interface.h
+++ b/src/stork/dap_client_interface.h
@@ -24,7 +24,7 @@
 #include "sock.h"
 
 #define WANT_CLASSAD_NAMESPACE
-#include "condor_fix_iostream.h"
+#include <iostream>
 #include "classad/classad_distribution.h"
 
 /** dap_submit()
diff --git a/src/stork/dap_logger.h b/src/stork/dap_logger.h
index 3620d89..e2bf14c 100644
--- a/src/stork/dap_logger.h
+++ b/src/stork/dap_logger.h
@@ -29,7 +29,7 @@
 #ifndef WANT_CLASSAD_NAMESPACE
 #define WANT_CLASSAD_NAMESPACE
 #endif
-#include "condor_fix_iostream.h"
+#include <iostream>
 #include "classad/classad_distribution.h"
 
 void write_dap_log(const char *logfilename, const char *status, const char *param1, const char *value1, const char *param2 = NULL, const char *value2 = NULL, const char *param3 = NULL, const char *value3 = NULL, const char *param4 = NULL, const char *value4 = NULL, const char *param5 = NULL, const char *value5 = NULL, const char *param6 = NULL, const char *value6 = NULL);
diff --git a/src/stork/dc_stork.h b/src/stork/dc_stork.h
index b087a5f..d0a29ce 100644
--- a/src/stork/dc_stork.h
+++ b/src/stork/dc_stork.h
@@ -21,7 +21,7 @@
 #define __DC_STORK_H__
 
 #define WANT_CLASSAD_NAMESPACE
-#include "condor_fix_iostream.h"
+#include <iostream>
 #include "classad/classad_distribution.h"
 
 class DCStork {
