diff --git a/src/condor_c++_util/classad_oldnew.cpp b/src/condor_c++_util/classad_oldnew.cpp
index d205e2a..607738c 100644
--- a/src/condor_c++_util/classad_oldnew.cpp
+++ b/src/condor_c++_util/classad_oldnew.cpp
@@ -145,6 +145,9 @@ getOldClassAdNoTypes( Stream *sock, classad::ClassAd& ad )
 	    free( secret_line );
         }
 
+		if ( strncmp( inputLine.Value(), "ConcurrencyLimit.", 17 ) == 0 ) {
+			inputLine.setChar( 16, '_' );
+		}
 		buffer += string(inputLine.Value()) + ";";
 	}
 	buffer += "]";
diff --git a/src/condor_negotiator.V6/Accountant.cpp b/src/condor_negotiator.V6/Accountant.cpp
index 9872da9..84fe736 100644
--- a/src/condor_negotiator.V6/Accountant.cpp
+++ b/src/condor_negotiator.V6/Accountant.cpp
@@ -1424,7 +1424,7 @@ void Accountant::ReportLimits(AttrList *attrList)
  	double count;
 	concurrencyLimits.startIterations();
 	while (concurrencyLimits.iterate(limit, count)) {
-		attr.sprintf("ConcurrencyLimit.%s = %f\n", limit.Value(), count);
+		attr.sprintf("ConcurrencyLimit_%s = %f\n", limit.Value(), count);
 		attrList->Insert(attr.Value());
 	}
 }
