From 0d7b6b12b08fd248028c3e260f0c7a304ba3e4b5 Mon Sep 17 00:00:00 2001
From: Lukas Slebodnik <slebodnik@sors.com>
Date: Thu, 26 May 2011 10:30:23 +0200
Subject: [PATCH] Fix return value of condor_q command

---
 src/condor_q.V6/queue.cpp |    6 ++++--
 1 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/condor_q.V6/queue.cpp b/src/condor_q.V6/queue.cpp
index 8316beb..fc4c51c 100644
--- a/src/condor_q.V6/queue.cpp
+++ b/src/condor_q.V6/queue.cpp
@@ -709,7 +709,8 @@ int main (int argc, char **argv)
 #ifdef HAVE_EXT_POSTGRESQL
 			case DIRECT_RDBMS:
 				if (useDB) {
-					if (sqfp(quillName, dbIpAddr, dbName, queryPassword, TRUE ))
+					if ( (retval = sqfp(quillName, dbIpAddr, dbName, 
+										queryPassword, TRUE) ) )
 					{
 						/* processed correctly, so do the next ad */
 						continue;
@@ -755,7 +756,8 @@ int main (int argc, char **argv)
 						didn't have a quill ad by that name. */
 
 					if((result2 == Q_OK) && quillAddr &&
-			   			sqfp(quillAddr, quillName, quillMachine, NULL, FALSE))
+			   			(retval = sqfp(quillAddr, quillName, quillMachine, 
+									   NULL, FALSE) ) )
 					{
 						/* processed correctly, so do the next ad */
 						continue;
-- 
1.7.4.1

