diff --git a/src/condor_q.V6/queue.cpp b/src/condor_q.V6/queue.cpp
index 1a6b748..c74f7ab 100644
--- a/src/condor_q.V6/queue.cpp
+++ b/src/condor_q.V6/queue.cpp
@@ -4285,8 +4285,14 @@ static bool read_userlog_file(const char *filename, ClassAdList &classads)
 	  jobClassAd->InsertAttr("EnteredCurrentStatus",eventTime);
 	  jobClassAd->InsertAttr("LastVacateTime",eventTime);
 	}
-
-	jobClassAd->InsertAttr("LastVacateReason",evict_event->getReason());
+	
+	{
+	  
+	  const char * reason=evict_event->getReason();
+	  if (reason!=NULL) {
+	    jobClassAd->InsertAttr("LastVacateReason",reason);
+	  }
+	}
 
 	// ignoring checkpointing and requeuing for now
 	// ignoring file transfer for now
@@ -4344,7 +4350,12 @@ static bool read_userlog_file(const char *filename, ClassAdList &classads)
 	  jobClassAd->InsertAttr("LastVacateTime",eventTime);
 	}
 
-	jobClassAd->InsertAttr("LastVacateReason",evict_event->getReason());
+	{
+	  const char * reason=evict_event->getReason();
+	  if (reason!=NULL) {
+	    jobClassAd->InsertAttr("LastVacateReason",reason);
+	  }
+	}
       }
       break;
       
@@ -4368,15 +4379,23 @@ static bool read_userlog_file(const char *filename, ClassAdList &classads)
 
 	// Update the time attributes
 	{
-	  condor_time_t eventTime = getEventTime(event);
-	  jobClassAd->InsertAttr("EnteredCurrentStatus",eventTime);
-	  if (prev_status!=IDLE) {
-	    jobClassAd->InsertAttr("LastVacateTime",eventTime);
-	    jobClassAd->InsertAttr("LastVacateReason",hold_event->getReason());
+	  const char * reason=hold_event->getReason();
+
+	  {
+	    condor_time_t eventTime = getEventTime(event);
+	    jobClassAd->InsertAttr("EnteredCurrentStatus",eventTime);
+	    if (prev_status!=IDLE) {
+	      jobClassAd->InsertAttr("LastVacateTime",eventTime);
+	      if (reason!=NULL) {
+		jobClassAd->InsertAttr("LastVacateReason",reason);
+	      }
+	    }
 	  }
-	}
 
-	jobClassAd->InsertAttr("HoldReason",hold_event->getReason());
+	  if (reason!=NULL) {
+	    jobClassAd->InsertAttr("HoldReason",reason);
+	  }
+	}
 	jobClassAd->InsertAttr("HoldReasonCode",hold_event->getReasonCode());
       	jobClassAd->InsertAttr("HoldReasonSubCode",hold_event->getReasonSubCode());
 	if (hold_event->getReasonCode()!=1) {
@@ -4410,7 +4429,12 @@ static bool read_userlog_file(const char *filename, ClassAdList &classads)
 
 	}
 
-	jobClassAd->InsertAttr("ReleaseReason",rel_event->getReason());
+	{
+	  const char * reason = rel_event->getReason();
+	  if (reason!=NULL) {
+	    jobClassAd->InsertAttr("ReleaseReason",reason);
+	  }
+	}
       }
       break;
 
@@ -4432,7 +4456,12 @@ static bool read_userlog_file(const char *filename, ClassAdList &classads)
 	  jobClassAd->InsertAttr("EnteredCurrentStatus",eventTime);
 	}
 
-	jobClassAd->InsertAttr("RemoveReason",abort_event->getReason());
+	{
+	  const char * reason = abort_event->getReason();
+	  if (reason!=NULL) {
+	    jobClassAd->InsertAttr("RemoveReason",reason);
+	  }
+	}
       }
       break;
 
