--- ticket.c.orig	2009-04-18 16:09:02.000000000 -0500
+++ ticket.c	2009-04-18 16:21:45.000000000 -0500
@@ -780,6 +780,10 @@
   azDrv = db_query(
     "SELECT tn,title FROM ticket WHERE derivedfrom=%d", tn);
   common_header("Ticket #%d", tn);
+  if( strcasecmp(az[17],"private")==0 && !g.okAdmin ) {
+	  @ <h2>This ticket is private.</h2>
+	  return;
+  }
   @ <h2>Ticket #%d(tn): %h(az[11])</h2>
   @ <blockquote>
   output_formatted(az[12], zPage);
@@ -1333,6 +1337,10 @@
     /* Don't stomp on someone elses edit if they hit submit first. */
     @ <input type="hidden" name="le" value="%d(le ? le : te)">
   }
+  if( strcasecmp(az[15],"private")==0 && !g.okAdmin ) {
+	  @ <h2>This ticket is private.</h2>
+	  return;
+  }
   @ <span class="nowrap">Ticket Number: %d(tn)</span><br>
   if( le && le < te ){
     @ <p class="error">
@@ -1518,6 +1526,7 @@
 void ticket_append(void){
   int tn, rn;
   char zPage[30];
+  char *zVis = NULL;
   int doPreview;
   int doSubmit;
   const char *zText;
@@ -1595,6 +1604,12 @@
   common_add_action_item( mprintf("tktview?tn=%h", zTn), "Cancel");
   common_header("Append Remarks To Ticket #%d", tn);
 
+  zVis = db_short_query("SELECT extra3 FROM ticket WHERE tn=%d", tn);
+  if( strcasecmp(zVis,"private")==0 && !g.okAdmin ) {
+	  @ <h2>This ticket is private.</h2>
+	  return;
+  }
+
   if( le && le < te ){
     @ <p class="error">
     @ Ticket #%d(tn) has been changed by someone else while you
@@ -1856,7 +1871,7 @@
   /* Get the record from the database.
   */
   db_add_functions();
-  az = db_query("SELECT title,origtime,owner FROM ticket WHERE tn=%d", tn);
+  az = db_query("SELECT title,origtime,owner,extra3 FROM ticket WHERE tn=%d", tn);
   if( az == NULL || az[0]==0 ){
     cgi_redirect("index");
     return;
@@ -1867,6 +1882,10 @@
   strftime(zDate, sizeof(zDate), "%Y-%b-%d %H:%M:%S", pTm);
 
   common_header("Ticket #%d History", tn);
+  if( strcasecmp(az[3],"private")==0 && !g.okAdmin ) {
+	  @ <h2>This ticket is private.</h2>
+	  return;
+  }
   @ <h2>Ticket %d(tn) History: %h(az[0])</h2>
   @ <ol>
   @ <li>Created %h(zDate) by %h(az[2])</li>
