From 8ad81b563866df25d7787363462a468341d7e0ae Mon Sep 17 00:00:00 2001
From: Peter MacKinnon <pmackinn@redhat.com>
Date: Wed, 18 Nov 2009 15:07:48 -0500
Subject: [PATCH] Fix line count logic in sshd.sh

---
 src/condor_scripts/sshd.sh |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/src/condor_scripts/sshd.sh b/src/condor_scripts/sshd.sh
index da1d0d4..ca749c7 100755
--- a/src/condor_scripts/sshd.sh
+++ b/src/condor_scripts/sshd.sh
@@ -130,7 +130,7 @@ then
 	do
 			/bin/rm -f contact
 			$CONDOR_CHIRP fetch $_CONDOR_REMOTE_SPOOL_DIR/contact $_CONDOR_SCRATCH_DIR/contact
-			lines=`wc -l $_CONDOR_SCRATCH_DIR/contact | awk "/$thisrun/"' {print $1}'`
+			lines=`grep -c $thisrun $_CONDOR_SCRATCH_DIR/contact`
 			if [ $lines -eq $_CONDOR_NPROCS ]
 			then
 				done=1
-- 
1.6.2.5

