diff --git a/src/condor_io/reli_sock.cpp b/src/condor_io/reli_sock.cpp
index 963526c..82ebb84 100644
--- a/src/condor_io/reli_sock.cpp
+++ b/src/condor_io/reli_sock.cpp
@@ -106,17 +106,11 @@ ReliSock::listen()
         return FALSE;
     }
 
-	// many modern OS's now support a >5 backlog, so we ask for 500,
-	// but since we don't know how they behave when you ask for too
-	// many, if 200 doesn't work we try progressively smaller numbers.
-	// you may ask, why not just use SOMAXCONN ?  unfortunately,
-	// it is not correct on several platforms such as Solaris, which
-	// accepts an unlimited number of socks but sets SOMAXCONN to 5.
-	if( ::listen( _sock, 500 ) < 0 ) {
-		if( ::listen( _sock, 300 ) < 0 ) 
-		if( ::listen( _sock, 200 ) < 0 ) 
-		if( ::listen( _sock, 100 ) < 0 ) 
-		if( ::listen( _sock, 5 ) < 0 ) {
+	// Ask for a (configurable) large backlog of connections. If this
+	// value is too large, the OS will cap it at the kernel's current
+	// maxiumum. Why not just use SOMAXCONN? Unfortunately, it's a
+	// fairly small value (128) on many platforms.
+	if( ::listen( _sock, param_integer( "SOCKET_LISTEN_BACKLOG", 500 ) ) < 0 ) {
 
             char const *self_address = get_sinful();
             if( !self_address ) {
@@ -130,7 +124,6 @@ ReliSock::listen()
 #endif
 
 			return FALSE;
-		}
 	}
 
 	dprintf( D_NETWORK, "LISTEN %s fd=%d\n", sock_to_string(_sock),
diff --git a/src/condor_io/shared_port_endpoint.cpp b/src/condor_io/shared_port_endpoint.cpp
index a778bf1..5b19178 100644
--- a/src/condor_io/shared_port_endpoint.cpp
+++ b/src/condor_io/shared_port_endpoint.cpp
@@ -337,7 +337,7 @@ SharedPortEndpoint::CreateListener()
 		return false;
 	}
 
-	if( listen(sock_fd,500) && listen(sock_fd,100) && listen(sock_fd,5) ) {
+	if( listen( sock_fd, param_integer( "SOCKET_LISTEN_BACKLOG", 500 ) ) ) {
 		dprintf(D_ALWAYS,
 				"ERROR: SharedPortEndpoint: failed to listen on %s: %s\n",
 				m_full_name.Value(), strerror(errno));
diff --git a/src/condor_utils/param_info.in b/src/condor_utils/param_info.in
index 95f643c..3e93a2a 100644
--- a/src/condor_utils/param_info.in
+++ b/src/condor_utils/param_info.in
@@ -298,6 +298,15 @@ customization=seldom
 friendly_name=Soap SSL Port
 review=?
 
+[SOCKET_LISTEN_BACKLOG]
+default=500
+range=1,
+type=int
+reconfig=false
+customization=seldom
+friendly_name=Socket Listen Backlog
+review=?
+
 [SOAP_SSL_PORT]
 default=0
 type=int
