--- condor-8.1.1/externals/bundles/cream/1.12.1_14/CMakeLists.txt.ORIG	2013-12-20 14:02:21.000000000 +0100
+++ condor-8.1.1/externals/bundles/cream/1.12.1_14/CMakeLists.txt	2014-01-15 10:44:14.000000000 +0100
@@ -46,7 +46,7 @@
 				    #--Patch step ----------
 				    PATCH_COMMAND patch -N -p1 -i ${CMAKE_CURRENT_SOURCE_DIR}/gsoap_cast.patch
 				    #--Configure step ----------
-				    CONFIGURE_COMMAND ./configure --prefix=${CREAM_INSTALL_LOC}
+				    CONFIGURE_COMMAND ./configure --prefix=${CREAM_INSTALL_LOC} --enable-ipv6
 				    #--Build Step ----------
 				    BUILD_COMMAND make
 				    BUILD_IN_SOURCE 1
@@ -160,7 +160,7 @@
 #########################################
 
 		# build gsoap plugin.
-		set (CSP_CC_FLAGS -g -O0 -W -Wall -Wno-unused-parameter -I. -I../interface -DWITH_NONAMESPACES -I${CREAM_INSTALL_LOC}/include -D_GNU_SOURCE)
+		set (CSP_CC_FLAGS -g -O0 -W -Wall -Wno-unused-parameter -I. -I../interface -DWITH_IPV6 -DWITH_NONAMESPACES -I${CREAM_INSTALL_LOC}/include -D_GNU_SOURCE)
 		condor_pre_external( CREAM_SOAP_PLUGIN org.glite.security.gsoap-plugin-2.0.1-3-p8 "lib;include" "done")
 		ExternalProject_Add(cream_soap_plugin
 					DEPENDS ${CREAMSOAP_REF} ${GLITE_SECURITY_REF}
