diff -rNup condor-7.6.2.orig/src/condor_utils/cron_job.cpp condor-7.6.2/src/condor_utils/cron_job.cpp
--- condor-7.6.2.orig/src/condor_utils/cron_job.cpp	2011-07-14 17:22:22.000000000 -0400
+++ condor-7.6.2/src/condor_utils/cron_job.cpp	2011-09-01 13:31:39.000000000 -0400
@@ -401,12 +401,11 @@ CronJob::Reaper( int exitPid, int exitSt
 				 GetName(), StateString() );
 		break;							// Do nothing
 
-		// Waiting for it to die...
+		// died either directly by term sig or kill sig, or other means while in sig code path.
 	case CRON_TERM_SENT:
 	case CRON_KILL_SENT:
-		break;							// Do nothing at all
-
-		// We've sent the process a signal, waiting for it to die
+		m_in_shutdown = false;
+		
 	default:
 		SetState( CRON_IDLE );			// Note that it's dead
 
@@ -428,11 +427,6 @@ CronJob::Reaper( int exitPid, int exitSt
 
 	}
 
-	// Note that we're dead
-	if ( IsInShutdown() ) {
-		SetState( CRON_DEAD );
-	}
-
 	// Process the output
 	ProcessOutputQueue( );
 
