diff --git a/src/condor_includes/condor_qmgr.h b/src/condor_includes/condor_qmgr.h
index 5e5012e..01330bc 100644
--- a/src/condor_includes/condor_qmgr.h
+++ b/src/condor_includes/condor_qmgr.h
@@ -346,4 +346,7 @@ int GetAllJobsByConstraint_Start( char const *constraint, char const *projection
 int GetAllJobsByConstraint_Next( ClassAd &ad );
 int GetJobQueuedCount();
 
+class StatisticsPool;
+void QmgmtAddStatsToPublish(StatisticsPool * ppool);
+
 #endif
diff --git a/src/condor_schedd.V6/qmgmt.cpp b/src/condor_schedd.V6/qmgmt.cpp
index 9cde502..443b2af 100644
--- a/src/condor_schedd.V6/qmgmt.cpp
+++ b/src/condor_schedd.V6/qmgmt.cpp
@@ -101,7 +101,11 @@ static time_t xact_start_time = 0;	// time at which the current transaction was
 static int cluster_initial_val = 1;		// first cluster number to use
 static int cluster_increment_val = 1;	// increment for cluster numbers of successive submissions 
 static int cluster_maximum_val = 0;     // maximum cluster id (default is 0, or 'no max')
+#if 0
 static int job_queued_count = 0;
+#else
+static stats_entry_recent<int> job_queued_count(0);
+#endif
 static Regex *queue_super_user_may_impersonate_regex = NULL;
 
 static void AddOwnerHistory(const MyString &user);
@@ -4960,5 +4964,13 @@ dirtyJobQueue()
 }
 
 int GetJobQueuedCount() {
-    return job_queued_count;
+    return job_queued_count.value;
+}
+
+void QmgmtAddStatsToPublish(StatisticsPool * ppool)
+{
+	ppool->AddProbe("JobsQueued", &job_queued_count, "JobsQueued", IF_VERBOSEPUB | job_queued_count.PubValue);
+	ppool->AddPublish("RecentJobsQueued", &job_queued_count, "RecentJobsQueued", IF_RECENTPUB | IF_VERBOSEPUB | job_queued_count.PubRecent);
+	//ppool->AddPublish("JobsQueuedDebug", &job_queued_count, "JobsQueuedDebug", IF_DEBUGPUB | IF_VERBOSEPUB | job_queued_count.PubDebug);
+
 }
diff --git a/src/condor_schedd.V6/schedd_stats.cpp b/src/condor_schedd.V6/schedd_stats.cpp
index 3ebcf80..e3debcf 100644
--- a/src/condor_schedd.V6/schedd_stats.cpp
+++ b/src/condor_schedd.V6/schedd_stats.cpp
@@ -21,6 +21,7 @@
 #include "condor_common.h"
 #include "scheduler.h"
 #include "schedd_stats.h"
+#include "condor_qmgr.h"
 #include "condor_config.h"
 #include "classad_helpers.h"
 
@@ -139,6 +140,7 @@ void ScheddStatistics::Init(int fOtherPool)
       SCHEDD_STATS_ADD_VAL(Pool, ShadowsRunning,               IF_BASICPUB);
       SCHEDD_STATS_PUB_PEAK(Pool, ShadowsRunning,              IF_BASICPUB);
 
+      QmgmtAddStatsToPublish(&Pool);
    //SCHEDD_STATS_PUB_DEBUG(Pool, JobsSubmitted,  IF_BASICPUB);
    //SCHEDD_STATS_PUB_DEBUG(Pool, JobsStarted,  IF_BASICPUB);
    //SCHEDD_STATS_PUB_DEBUG(Pool, JobsCompleted,  IF_BASICPUB);
@@ -396,9 +398,6 @@ void ScheddOtherStatsMgr::DeferJobsSubmitted(int cluster, int proc)
 	}
 }
 
-// from condor_qmgr.h...
-extern ClassAd *GetJobAd(int cluster_id, int proc_id, bool expStardAttrs = false, bool persist_expansions = true );
-extern void FreeJobAd(ClassAd *&ad);
 
 // finish deferred counting of submitted jobs.
 void ScheddOtherStatsMgr::CountJobsSubmitted()
