diff --git a/src/condor_utils/CMakeLists.txt b/src/condor_utils/CMakeLists.txt
index d5beaf6..246f097 100644
--- a/src/condor_utils/CMakeLists.txt
+++ b/src/condor_utils/CMakeLists.txt
@@ -87,6 +87,13 @@ endif()
 
 target_link_libraries(condorapi ${CLASSADS_FOUND};${APILIBDEP})
 
+if (LINUX OR DARWIN)
+    condor_shared_lib( condorapi_shared "${ApiSrcs}" )
+    add_dependencies( condorapi_shared condor_utils )
+    target_link_libraries(condorapi_shared ${CLASSADS_FOUND};${APILIBDEP})
+    set_target_properties( condorapi_shared PROPERTIES OUTPUT_NAME condorapi )
+endif()
+
 if (WANT_FULL_DEPLOYMENT)
 	install (FILES MyString.h
 			 condor_event.h
@@ -102,6 +109,9 @@ if (WANT_FULL_DEPLOYMENT)
 			DESTINATION ${C_INCLUDE})
 
     install ( TARGETS condorapi DESTINATION ${C_LIB} )
+    if (LINUX OR DARWIN)
+        install ( TARGETS condorapi_shared DESTINATION ${C_LIB} )
+    endif()
 endif(WANT_FULL_DEPLOYMENT)
 
 add_definitions(-DENABLE_STATE_DUMP)
