diff --git a/src/condor_gridmanager/globusjob.cpp b/src/condor_gridmanager/globusjob.cpp
index 585612c..6b62b28 100644
--- a/src/condor_gridmanager/globusjob.cpp
+++ b/src/condor_gridmanager/globusjob.cpp
@@ -3541,11 +3541,14 @@ GlobusJob::JmShouldSleep()
 		return false;
 	}
 	if ( jmProxyExpireTime < jobProxy->expiration_time ) {
-		if ( time(NULL) >= jmProxyExpireTime - 6*3600 ) {
+		// Don't forward the refreshed proxy if the remote proxy has more
+		// than GRIDMANAGER_PROXY_RENEW_LIMIT time left.
+		int renew_min = param_integer( "GRIDMANAGER_PROXY_RENEW_MINIMUM", 6*3600 );
+		if ( time(NULL) >= jmProxyExpireTime - renew_min ) {
 			return false;
 		} else {
 			daemonCore->Reset_Timer( evaluateStateTid,
-								 ( jmProxyExpireTime - 6*3600 ) - time(NULL) );
+								 ( jmProxyExpireTime - renew_min ) - time(NULL) );
 		}
 	}
 	if ( condorState != IDLE && condorState != RUNNING ) {
