--- condor-8.1.1/src/nordugrid_gahp/globus_ftp_client_attr.c.ORIG	2013-12-20 14:02:49.000000000 +0100
+++ condor-8.1.1/src/nordugrid_gahp/globus_ftp_client_attr.c	2014-02-25 15:03:59.000000000 +0100
@@ -990,7 +990,7 @@
     i_attr->read_all_intermediate_callback= GLOBUS_NULL;
     i_attr->read_all_intermediate_callback_arg
 					= GLOBUS_NULL;
-    i_attr->allow_ipv6                  = GLOBUS_FALSE;
+    i_attr->allow_ipv6                  = (getenv("GLOBUS_FTP_CLIENT_IPV6") ? GLOBUS_TRUE : GLOBUS_FALSE);
     i_attr->resume_third_party		= GLOBUS_FALSE;
     i_attr->force_striped		= GLOBUS_FALSE;
     i_attr->allocated_size		= 0;
