diff --git a/src/condor_c++_util/uids.cpp b/src/condor_c++_util/uids.cpp
index 6c06201..30054b4 100644
--- a/src/condor_c++_util/uids.cpp
+++ b/src/condor_c++_util/uids.cpp
@@ -534,9 +534,16 @@ _set_priv(priv_state s, const char *file, int line, int dologging)
 				}
 				ImpersonateLoggedOnUser(CurrUserHandle);
 			} else {
-				// Tried to set_user_priv() but it failed, so bail out!
-
-				EXCEPT("set_user_priv() failed!");
+				// we don't want to exit here because reusing the shadow in 7.5.4
+				// ends up here because of a dprintf inside uninit_user_ids
+				if ( ! UserIdsInited) {
+					if (dologging) {
+						dprintf(D_ALWAYS, "set_user_priv() called when UserIds not inited!\n");
+					}
+				}  else {
+					// Tried to set_user_priv() but it failed, so log and continue.
+					EXCEPT("set_user_priv() failed!");
+				}
 			}
 			break;
 		case PRIV_UNKNOWN:		/* silently ignore */
